{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Introduction\n",
    "\n",
    "This notebook demonstrates basic usage of BioThings Explorer, an engine for autonomously querying a distributed knowledge graph of biomedical knowledge.\n",
    "\n",
    "BioThings Explorer can answer two classes of queries -- \"EXPLAIN\" and \"PREDICT\". EXPLAIN queries are described in [EXPLAIN_demo.ipynb](EXPLAIN_demo.ipynb). Here, we describe PREDICT queries and how to use BioThings Explorer to execute them. A more detailed overview of the BioThings Explorer systems is provided in [these slides](https://docs.google.com/presentation/d/1QWQqqQhPD_pzKryh6Wijm4YQswv8pAjleVORCPyJyDE/edit?usp=sharing).\n",
    "\n",
    "EXPLAIN queries are designed to **identify plausible reasoning chains to explain the relationship between two specific biomedical entities (objects or concepts)**.  For example, in this notebook, we explore the question:\n",
    "\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;\"*Why does imatinib have an effect on the treatment of chronic myelogenous leukemia (CML)?*\"  \n",
    "\n",
    "Later, we also compare those results to a similar query looking at imatinib's role in treating gastrointestinal stromal tumors (GIST).\n",
    "\n",
    "**To experiment with an executable version of this notebook, [load it in Google Colaboratory](https://colab.research.google.com/github/biothings/biothings_explorer/blob/master/jupyter%20notebooks/EXPLAIN_demo.ipynb).**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 0: Load BioThings Explorer modules"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First, install the `biothings_explorer` and `biothings_schema` packages, as described in this [README](https://github.com/biothings/biothings_explorer/blob/master/jupyter%20notebooks/README.md#prerequisite).  This only needs to be done once (but it's included here for compability with [Colab](https://colab.research.google.com/))."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%capture\n",
    "!pip install git+https://github.com/biothings/biothings_explorer#egg=biothings_explorer"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, import the relevant modules:\n",
    "\n",
    "* **Hint**: Find corresponding BioThings representations to use in BioThings Explorer based on user input (could be any database IDs, symbols, names)\n",
    "* **FindConnection**: Find the relationship(s) between a specific entity and an entity type (PREDICT) or two specific entities (EXPLAIN))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import modules from biothings_explorer\n",
    "from biothings_explorer.hint import Hint\n",
    "from biothings_explorer.user_query_dispatcher import FindConnection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def hint_display(query, hint_result):\n",
    "    \"\"\"\n",
    "    input: query (string used in hint query), hint_result (object returned from hint query)\n",
    "    hint_result is a dictionary of lists of dictionaries\n",
    "    returns: None\n",
    "    show the name and type of all results returned by the query\n",
    "    \"\"\"\n",
    "    display = ['name', 'type']  ## replace with the parts of the BioThings object you want to see\n",
    "    concise_results = []\n",
    "    \n",
    "    for BT_type, result in hint_result.items():\n",
    "        if result:  ## basically if it's not empty\n",
    "            for items in result:\n",
    "                concise_results.append([items[ele] for ele in display])\n",
    "    \n",
    "    print('There are {total} BioThings objects returned for {ht}:'.format(\\\n",
    "                total = len(concise_results), ht = query))\n",
    "    for (idx, display_info) in enumerate(concise_results):\n",
    "        print('{i}: {val}'.format(i = idx, val = display_info))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 1: Find representations of \"chronic myelogenous leukemia\" and \"imatinib\" in BTE"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this step, BioThings Explorer translates our query strings \"chronic myelogenous leukemia\" and \"imatinib\" into BioThings objects, which contain mappings to many common identifiers.  Generally, the top result returned by the `Hint` module will match what you want, but you should confirm that using the identifiers shown. \n",
    "\n",
    "BioThings types correspond to children and descendants of [BiologicalEntity](https://biolink.github.io/biolink-model/docs/BiologicalEntity.html) from the [Biolink Model](https://biolink.github.io/biolink-model/docs/), including `Disease` (e.g., \"lupus\"), `ChemicalSubstance` (e.g., \"acetaminophen\"), `Gene` (e.g., \"CDK2\"), `BiologicalProcess` (e.g., \"T cell differentiation\"), and `Pathway` (e.g., \"Citric acid cycle\"). **However, [only a subset of the Biolink BiologicalEntity children / descendants are currently implemented in BTE](https://smart-api.info/portal/translator/metakg)**. More biomedical object types will be available as more knowledge sources (APIs) are added to the system. \n",
    "\n",
    "**Note that the type `BiologicalEntity` means any BioThings type currently implemented in BTE will be accepted.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 2 BioThings objects returned for chronic myelogenous leukemia:\n",
      "0: ['chronic myelogenous leukemia, BCR-ABL1 positive', 'Disease']\n",
      "1: ['blast phase chronic myelogenous leukemia, BCR-ABL1 positive', 'Disease']\n"
     ]
    }
   ],
   "source": [
    "ht = Hint()\n",
    "cml_str = \"chronic myelogenous leukemia\"\n",
    "# find all potential representations of CML\n",
    "cml_hint = ht.query(cml_str)\n",
    "\n",
    "hint_display(cml_str, cml_hint)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Based on the information above, we'll pick the top Disease choice (indexed at 0) for CML for our query. We can look at identifier mappings inside this BioThings object. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'MONDO': 'MONDO:0011996',\n",
       " 'DOID': 'DOID:8552',\n",
       " 'UMLS': 'C0023473',\n",
       " 'name': 'chronic myelogenous leukemia, BCR-ABL1 positive',\n",
       " 'OMIM': '608232',\n",
       " 'ORPHANET': '521',\n",
       " 'primary': {'identifier': 'MONDO',\n",
       "  'cls': 'Disease',\n",
       "  'value': 'MONDO:0011996'},\n",
       " 'display': 'MONDO(MONDO:0011996) DOID(DOID:8552) OMIM(608232) ORPHANET(521) UMLS(C0023473) name(chronic myelogenous leukemia, BCR-ABL1 positive)',\n",
       " 'type': 'Disease'}"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# select the first Disease object for our CML query\n",
    "cml = cml_hint['Disease'][0]\n",
    "cml"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 10 BioThings objects returned for imatinib:\n",
      "0: ['imatinib', 'ChemicalSubstance']\n",
      "1: ['imatinib methanesulfonate', 'ChemicalSubstance']\n",
      "2: ['linkable imatinib analogue', 'ChemicalSubstance']\n",
      "3: ['IMATINIB', 'ChemicalSubstance']\n",
      "4: ['IMATINIB MESYLATE', 'ChemicalSubstance']\n",
      "5: ['HYPEREOSINOPHILIC SYNDROME, IDIOPATHIC, RESISTANT TO IMATINIB', 'Disease']\n",
      "6: ['CHRONIC MYELOID LEUKEMIA, RESISTANT TO IMATINIB', 'Disease']\n",
      "7: ['LEUKEMIA, PHILADELPHIA CHROMOSOME-POSITIVE, RESISTANT TO IMATINIB', 'Disease']\n",
      "8: ['Imatinib Pathway, Pharmacokinetics/Pharmacodynamics', 'Pathway']\n",
      "9: ['Imatinib and Chronic Myeloid Leukemia', 'Pathway']\n"
     ]
    }
   ],
   "source": [
    "# find all potential representations of imatinib\n",
    "imatinib_str = \"imatinib\"\n",
    "imatinib_hint = ht.query(imatinib_str)\n",
    "\n",
    "hint_display(imatinib_str, imatinib_hint)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "While multiple results look valid (indexed at 0, 1, 3, 4), for now we'll pick the top ChemicalSubstance choice (indexed at 0) for imatinib for our query. We can look at identifier mappings inside this BioThings object. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'DRUGBANK': 'DB00619',\n",
       " 'CHEBI': 'CHEBI:45783',\n",
       " 'name': 'imatinib',\n",
       " 'primary': {'identifier': 'CHEBI',\n",
       "  'cls': 'ChemicalSubstance',\n",
       "  'value': 'CHEBI:45783'},\n",
       " 'display': 'CHEBI(CHEBI:45783) DRUGBANK(DB00619) name(imatinib)',\n",
       " 'type': 'ChemicalSubstance'}"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# select the correct representation of imatinib\n",
    "imatinib = imatinib_hint['ChemicalSubstance'][0]\n",
    "imatinib"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 2: Find intermediate nodes connecting imatinib and chronic myelogenous leukemia"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this section, we find some paths in the knowledge graph that connect imatinib and chronic myelogenous leukemia.  To do that, we will use `FindConnection`.  This class is a convenient wrapper around two advanced functions for **query path planning** and **query path execution**. More advanced features for both query path planning and query path execution are in development and will be documented in the coming months. \n",
    "\n",
    "The parameters for `FindConnection` are described below:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on function __init__ in module biothings_explorer.user_query_dispatcher:\n",
      "\n",
      "__init__(self, input_obj, output_obj, intermediate_nodes, registry=None)\n",
      "    Find relationships in the Knowledge Graph between an Input Object and an Output Object.\n",
      "    \n",
      "    Args:\n",
      "        input_obj (required): must be an object returned from Hint corresponding to a specific biomedical entity.\n",
      "                            Examples:\n",
      "                Hint().query(\"Fanconi anemia\")['DiseaseOrPhenotypicFeature'][0]\n",
      "                Hint().query(\"acetaminophen\")['ChemicalSubstance'][0]\n",
      "    \n",
      "        output_obj (required): must EITHER be an object returned from Hint corresponding to a specific biomedical\n",
      "                            entity, OR be a string or list of strings corresponding to Biolink Entity classes.\n",
      "                            Examples:\n",
      "                Hint().query(\"acetaminophen\")['ChemicalSubstance'][0]\n",
      "                'Gene'\n",
      "                ['Gene','ChemicalSubstance']\n",
      "    \n",
      "        intermediate_nodes (required): the semantic type(s) of the intermediate node(s).  Examples:\n",
      "                None                         : no intermediate node, find direct connections only\n",
      "                []                           : no intermediate node, find direct connections only\n",
      "                ['BiologicalEntity']         : one intermediate node of any semantic type\n",
      "                ['Gene']                     : one intermediate node that must be a Gene\n",
      "                [('Gene','Pathway')]         : one intermediate node that must be a Gene or a Pathway\n",
      "                ['Gene','Pathway']           : two intermediate nodes, first must be a Gene, second must be a Pathway.\n",
      "                ['Gene',('Pathway','Gene')]  : two intermediate nodes, first must be a Gene, second must be a Pathway or Gene.\n",
      "                                                **NOTE**: queries with more than one intermediate node are currently not supported\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(FindConnection.__init__)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here, we formulate a `FindConnection` query with CML as the `input_obj`, imatinib as the `output_obj`.  We further specify with the `intermediate_nodes` parameter that we are looking for paths joining chronic myelogenous leukemia and imatinib with *one* intermediate node that is a \"Gene\" (BioThings type).  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "fc = FindConnection(input_obj = cml, output_obj = imatinib, \\\n",
    "                    intermediate_nodes = 'Gene')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We next execute the `connect` method, which performs the **query path planning** and **query path execution** process.  In short, BioThings Explorer is deconstructing the query into individual API calls, executing those API calls, then assembling the results.\n",
    "\n",
    "A verbose log of this process is displayed below:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==========\n",
      "========== QUERY PARAMETER SUMMARY ==========\n",
      "==========\n",
      "\n",
      "BTE will find paths that join 'chronic myelogenous leukemia, BCR-ABL1 positive' and 'imatinib'. Paths will have 1 intermediate node.\n",
      "\n",
      "Intermediate node #1 will have these type constraints: Gene\n",
      "\n",
      "\n",
      "==========\n",
      "========== QUERY #1 -- fetch all Gene entities linked to 'chronic myelogenous leukemia, BCR-ABL1 positive' ==========\n",
      "==========\n",
      "\n",
      "==== Step #1: Query path planning ====\n",
      "\n",
      "Because chronic myelogenous leukemia, BCR-ABL1 positive is of type 'Disease', BTE will query our meta-KG for APIs that can take 'Disease' as input and 'Gene' as output\n",
      "\n",
      "BTE found 10 apis:\n",
      "\n",
      "API 1. pharos(1 API call)\n",
      "API 2. scigraph(1 API call)\n",
      "API 3. cord_disease(1 API call)\n",
      "API 4. semmed_disease(15 API calls)\n",
      "API 5. hetio(1 API call)\n",
      "API 6. DISEASES(1 API call)\n",
      "API 7. mgi_gene2phenotype(1 API call)\n",
      "API 8. mydisease(1 API call)\n",
      "API 9. biolink(1 API call)\n",
      "API 10. scibite(1 API call)\n",
      "\n",
      "\n",
      "==== Step #2: Query path execution ====\n",
      "NOTE: API requests are dispatched in parallel, so the list of APIs below is ordered by query time.\n",
      "\n",
      "API 7.1: https://mydisease.info/v1/query?fields=disgenet.genes_related_to_disease.gene_id (POST -d q=C0023474,C0023473&scopes=mondo.xrefs.umls, disgenet.xrefs.umls)\n",
      "API 8.1: https://pending.biothings.io/mgigene2phenotype/query?fields=_id&size=300 (POST -d q=DOID:8552&scopes=mgi.associated_with_disease.doid)\n",
      "API 6.1: https://pending.biothings.io/DISEASES/query?fields=DISEASES.associatedWith (POST -d q=DOID:8552&scopes=DISEASES.doid)\n",
      "API 4.1: https://biothings.ncats.io/semmed/query?fields=affects (POST -d q=C0023474,C0023473&scopes=umls)\n",
      "API 4.7: https://biothings.ncats.io/semmed/query?fields=positively_regulated_by (POST -d q=C0023474,C0023473&scopes=umls)\n",
      "API 4.3: https://biothings.ncats.io/semmed/query?fields=prevented_by (POST -d q=C0023474,C0023473&scopes=umls)\n",
      "API 9.1: https://api.monarchinitiative.org/api/bioentity/disease/MONDO:0011996/genes?rows=200\n",
      "API 3.1: https://biothings.ncats.io/cord_disease/query?fields=associated_with (POST -d q=DOID:8552&scopes=doid)\n",
      "API 4.11: https://biothings.ncats.io/semmed/query?fields=negatively_regulates (POST -d q=C0023474,C0023473&scopes=umls)\n",
      "API 4.5: https://biothings.ncats.io/semmed/query?fields=physically_interacts_with (POST -d q=C0023474,C0023473&scopes=umls)\n",
      "API 4.10: https://biothings.ncats.io/semmed/query?fields=affected_by (POST -d q=C0023474,C0023473&scopes=umls)\n",
      "API 4.15: https://biothings.ncats.io/semmed/query?fields=negatively_regulated_by (POST -d q=C0023474,C0023473&scopes=umls)\n",
      "API 4.8: https://biothings.ncats.io/semmed/query?fields=derives_from (POST -d q=C0023474,C0023473&scopes=umls)\n",
      "API 4.4: https://biothings.ncats.io/semmed/query?fields=positively_regulates (POST -d q=C0023474,C0023473&scopes=umls)\n",
      "API 4.13: https://biothings.ncats.io/semmed/query?fields=disrupts (POST -d q=C0023474,C0023473&scopes=umls)\n",
      "API 4.9: https://biothings.ncats.io/semmed/query?fields=disrupted_by (POST -d q=C0023474,C0023473&scopes=umls)\n",
      "API 4.12: https://biothings.ncats.io/semmed/query?fields=related_to (POST -d q=C0023474,C0023473&scopes=umls)\n",
      "API 4.14: https://biothings.ncats.io/semmed/query?fields=caused_by (POST -d q=C0023474,C0023473&scopes=umls)\n",
      "API 4.2: https://biothings.ncats.io/semmed/query?fields=coexists_with (POST -d q=C0023474,C0023473&scopes=umls)\n",
      "API 4.6: https://biothings.ncats.io/semmed/query?fields=treated_by (POST -d q=C0023474,C0023473&scopes=umls)\n",
      "API 10.1: https://automat.renci.org/cord19_scibite_v2/disease/gene/MONDO:0011996\n",
      "API 10.1 scibite failed\n",
      "API 2.1: https://automat.renci.org/cord19_scigraph_v2/disease/gene/MONDO:0011996\n",
      "API 2.1 scigraph failed\n",
      "API 1.1: https://automat.renci.org/pharos/disease/gene/MONDO:0011996\n",
      "API 5.1: https://automat.renci.org/hetio/disease/gene/MONDO:0011996\n",
      "\n",
      "\n",
      "==== Step #3: Output normalization ====\n",
      "\n",
      "API 8.1 mgi_gene2phenotype: 8 hits\n",
      "API 4.1 semmed_disease: No hits\n",
      "API 4.2 semmed_disease: No hits\n",
      "API 6.1 DISEASES: 3 hits\n",
      "API 4.3 semmed_disease: 7 hits\n",
      "API 4.4 semmed_disease: No hits\n",
      "API 4.5 semmed_disease: No hits\n",
      "API 4.6 semmed_disease: 136 hits\n",
      "API 4.7 semmed_disease: No hits\n",
      "API 1.1 pharos: 7 hits\n",
      "API 4.8 semmed_disease: No hits\n",
      "API 4.9 semmed_disease: 26 hits\n",
      "API 3.1 cord_disease: 146 hits\n",
      "API 4.10 semmed_disease: 64 hits\n",
      "API 5.1 hetio: No hits\n",
      "API 4.11 semmed_disease: No hits\n",
      "API 4.12 semmed_disease: 557 hits\n",
      "API 7.1 mydisease: No hits\n",
      "API 4.13 semmed_disease: No hits\n",
      "API 4.14 semmed_disease: 92 hits\n",
      "API 9.1 biolink: 4 hits\n",
      "API 4.15 semmed_disease: No hits\n",
      "\n",
      "After id-to-object translation, BTE retrieved 740 unique objects.\n",
      "\n",
      "\n",
      "==========\n",
      "========== QUERY #2 -- fetch all Gene entities linked to 'imatinib' ==========\n",
      "==========\n",
      "\n",
      "==== Step #1: Query path planning ====\n",
      "\n",
      "Because imatinib is of type 'ChemicalSubstance', BTE will query our meta-KG for APIs that can take 'ChemicalSubstance' as input and 'Gene' as output\n",
      "\n",
      "BTE found 10 apis:\n",
      "\n",
      "API 1. cord_chemical(1 API call)\n",
      "API 2. pharos(1 API call)\n",
      "API 3. scigraph(1 API call)\n",
      "API 4. mychem(3 API calls)\n",
      "API 5. chembio(1 API call)\n",
      "API 6. ctd(2 API calls)\n",
      "API 7. semmed_chemical(13 API calls)\n",
      "API 8. dgidb(2 API calls)\n",
      "API 9. hmdb(1 API call)\n",
      "API 10. scibite(1 API call)\n",
      "\n",
      "\n",
      "==== Step #2: Query path execution ====\n",
      "NOTE: API requests are dispatched in parallel, so the list of APIs below is ordered by query time.\n",
      "\n",
      "API 7.2: https://biothings.ncats.io/semmedchemical/query?fields=positively_regulates (POST -d q=C0935989&scopes=umls)\n",
      "API 7.5: https://biothings.ncats.io/semmedchemical/query?fields=related_to (POST -d q=C0935989&scopes=umls)\n",
      "API 7.8: https://biothings.ncats.io/semmedchemical/query?fields=produces (POST -d q=C0935989&scopes=umls)\n",
      "API 1.1: https://biothings.ncats.io/cord_chemical/query?fields=associated_with (POST -d q=CHEBI:45783&scopes=chebi)\n",
      "API 6.3: https://mychem.info/v1/query?fields=drugbank.enzymes (POST -d q=DB00619&scopes=drugbank.id)\n",
      "API 7.3: https://biothings.ncats.io/semmedchemical/query?fields=negatively_regulated_by (POST -d q=C0935989&scopes=umls)\n",
      "API 6.2: https://mychem.info/v1/query?fields=drugcentral.bioactivity (POST -d q=CHEMBL1642,CHEMBL941&scopes=chembl.molecule_chembl_id)\n",
      "API 7.6: https://biothings.ncats.io/semmedchemical/query?fields=disrupts (POST -d q=C0935989&scopes=umls)\n",
      "API 6.1: https://mychem.info/v1/query?fields=drugbank.targets (POST -d q=DB00619&scopes=drugbank.id)\n",
      "API 7.7: https://biothings.ncats.io/semmedchemical/query?fields=coexists_with (POST -d q=C0935989&scopes=umls)\n",
      "API 7.10: https://biothings.ncats.io/semmedchemical/query?fields=negatively_regulates (POST -d q=C0935989&scopes=umls)\n",
      "API 7.12: https://biothings.ncats.io/semmedchemical/query?fields=affected_by (POST -d q=C0935989&scopes=umls)\n",
      "API 7.4: https://biothings.ncats.io/semmedchemical/query?fields=disrupted_by (POST -d q=C0935989&scopes=umls)\n",
      "API 7.13: https://biothings.ncats.io/semmedchemical/query?fields=positively_regulated_by (POST -d q=C0935989&scopes=umls)\n",
      "API 7.11: https://biothings.ncats.io/semmedchemical/query?fields=affects (POST -d q=C0935989&scopes=umls)\n",
      "API 7.9: https://biothings.ncats.io/semmedchemical/query?fields=produced_by (POST -d q=C0935989&scopes=umls)\n",
      "API 7.1: https://biothings.ncats.io/semmedchemical/query?fields=physically_interacts_with (POST -d q=C0935989&scopes=umls)\n",
      "API 8.1: http://dgidb.genome.wustl.edu/api/v2/interactions.json?drugs=CHEMBL1642\n",
      "API 8.2: http://dgidb.genome.wustl.edu/api/v2/interactions.json?drugs=CHEMBL941\n",
      "API 4.2: http://ctdbase.org/tools/batchQuery.go?inputType=chem&report=genes_curated&format=json&inputTerms=C097613\n",
      "API 4.1: http://ctdbase.org/tools/batchQuery.go?inputType=chem&report=genes_curated&format=json&inputTerms=D000068877\n",
      "API 10.1: https://automat.renci.org/cord19_scibite_v2/chemical_substance/gene/CHEBI:45783\n",
      "API 10.1 scibite failed\n",
      "API 3.1: https://automat.renci.org/cord19_scigraph_v2/chemical_substance/gene/CHEBI:45783\n",
      "API 3.1 scigraph failed\n",
      "API 2.1: https://automat.renci.org/pharos/chemical_substance/gene/CHEBI:45783\n",
      "API 5.1: https://automat.renci.org/chembio/chemical_substance/gene/CHEBI:45783\n",
      "API 9.1: https://automat.renci.org/hmdb/chemical_substance/gene/CHEBI:45783\n",
      "\n",
      "\n",
      "==== Step #3: Output normalization ====\n",
      "\n",
      "API 7.1 semmed_chemical: 226 hits\n",
      "API 5.1 chembio: No hits\n",
      "API 6.1 mychem: 9 hits\n",
      "API 6.2 mychem: 80 hits\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "API 6.3 mychem: 9 hits\n",
      "API 2.1 pharos: 6 hits\n",
      "API 7.2 semmed_chemical: 96 hits\n",
      "API 7.3 semmed_chemical: 36 hits\n",
      "API 7.4 semmed_chemical: No hits\n",
      "API 7.5 semmed_chemical: No hits\n",
      "API 8.1 dgidb: 5 hits\n",
      "API 8.2 dgidb: 34 hits\n",
      "API 9.1 hmdb: No hits\n",
      "API 7.6 semmed_chemical: No hits\n",
      "API 7.7 semmed_chemical: 119 hits\n",
      "API 7.8 semmed_chemical: No hits\n",
      "API 7.9 semmed_chemical: 11 hits\n",
      "API 1.1 cord_chemical: 172 hits\n",
      "API 4.1 ctd: 199 hits\n",
      "API 4.2 ctd: No hits\n",
      "API 7.10 semmed_chemical: 166 hits\n",
      "API 7.11 semmed_chemical: No hits\n",
      "API 7.12 semmed_chemical: No hits\n",
      "API 7.13 semmed_chemical: 39 hits\n",
      "\n",
      "After id-to-object translation, BTE retrieved 728 unique objects.\n",
      "\n",
      "==========\n",
      "========== Final assembly of results ==========\n",
      "==========\n",
      "\n",
      "\n",
      "BTE found 249 unique intermediate nodes connecting 'chronic myelogenous leukemia, BCR-ABL1 positive' and 'imatinib'\n"
     ]
    }
   ],
   "source": [
    "# set verbose=True will display all steps which BTE takes to find the connection\n",
    "fc.connect(verbose = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 3: Display and filter results"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This section demonstrates post-query filtering done in Python. Later, more advanced filtering functions will be added to the **query path execution** module for interleaved filtering, thereby enabling longer query paths. More details to come...\n",
    "\n",
    "First, results can be exported to a data frame. Let's examine a sample of those results."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1028, 16)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>input</th>\n",
       "      <th>input_type</th>\n",
       "      <th>pred1</th>\n",
       "      <th>pred1_source</th>\n",
       "      <th>pred1_api</th>\n",
       "      <th>pred1_pubmed</th>\n",
       "      <th>node1_type</th>\n",
       "      <th>node1_name</th>\n",
       "      <th>node1_id</th>\n",
       "      <th>pred2</th>\n",
       "      <th>pred2_source</th>\n",
       "      <th>pred2_api</th>\n",
       "      <th>pred2_pubmed</th>\n",
       "      <th>output_type</th>\n",
       "      <th>output_name</th>\n",
       "      <th>output_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1319</th>\n",
       "      <td>CHRONIC MYELOGENOUS LEUKEMIA, BCR-ABL1 POSITIVE</td>\n",
       "      <td>Disease</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Disease API</td>\n",
       "      <td>27573699</td>\n",
       "      <td>Gene</td>\n",
       "      <td>MYC</td>\n",
       "      <td>NCBIGene:4609</td>\n",
       "      <td>positively_regulated_by</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Chemical API</td>\n",
       "      <td>22842456</td>\n",
       "      <td>Gene</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>name:IMATINIB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>563</th>\n",
       "      <td>CHRONIC MYELOGENOUS LEUKEMIA, BCR-ABL1 POSITIVE</td>\n",
       "      <td>Disease</td>\n",
       "      <td>related_to</td>\n",
       "      <td>None</td>\n",
       "      <td>BioLink API</td>\n",
       "      <td>None</td>\n",
       "      <td>Gene</td>\n",
       "      <td>BCR</td>\n",
       "      <td>NCBIGene:613</td>\n",
       "      <td>positively_regulates</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Chemical API</td>\n",
       "      <td>20228846,22204395,22641215</td>\n",
       "      <td>Gene</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>name:IMATINIB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1092</th>\n",
       "      <td>CHRONIC MYELOGENOUS LEUKEMIA, BCR-ABL1 POSITIVE</td>\n",
       "      <td>Disease</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Disease API</td>\n",
       "      <td>10092207,10224280,10490629,10676660,10833515,1...</td>\n",
       "      <td>Gene</td>\n",
       "      <td>MTTP</td>\n",
       "      <td>NCBIGene:4547</td>\n",
       "      <td>coexists_with</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Chemical API</td>\n",
       "      <td>12669727,15283151,21505592,22713161,23434731</td>\n",
       "      <td>Gene</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>name:IMATINIB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1537</th>\n",
       "      <td>CHRONIC MYELOGENOUS LEUKEMIA, BCR-ABL1 POSITIVE</td>\n",
       "      <td>Disease</td>\n",
       "      <td>related_to</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Disease API</td>\n",
       "      <td>10688886</td>\n",
       "      <td>Gene</td>\n",
       "      <td>SQSTM1</td>\n",
       "      <td>NCBIGene:8878</td>\n",
       "      <td>related_to</td>\n",
       "      <td>CTD</td>\n",
       "      <td>CTD API</td>\n",
       "      <td>22641616,30165626</td>\n",
       "      <td>Gene</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>name:IMATINIB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>700</th>\n",
       "      <td>CHRONIC MYELOGENOUS LEUKEMIA, BCR-ABL1 POSITIVE</td>\n",
       "      <td>Disease</td>\n",
       "      <td>treated_by</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Disease API</td>\n",
       "      <td>15717993,16205964,16596277,17694264,17891251,2...</td>\n",
       "      <td>Gene</td>\n",
       "      <td>KIT</td>\n",
       "      <td>NCBIGene:3815</td>\n",
       "      <td>positively_regulates</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Chemical API</td>\n",
       "      <td>17139461,28765927,28878021</td>\n",
       "      <td>Gene</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>name:IMATINIB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>911</th>\n",
       "      <td>CHRONIC MYELOGENOUS LEUKEMIA, BCR-ABL1 POSITIVE</td>\n",
       "      <td>Disease</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Disease API</td>\n",
       "      <td>22823957</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ABCB1</td>\n",
       "      <td>NCBIGene:5243</td>\n",
       "      <td>physically_interacts_with</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Chemical API</td>\n",
       "      <td>16157201,17495881,17690695,18669873,19225039,2...</td>\n",
       "      <td>Gene</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>name:IMATINIB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1264</th>\n",
       "      <td>CHRONIC MYELOGENOUS LEUKEMIA, BCR-ABL1 POSITIVE</td>\n",
       "      <td>Disease</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Disease API</td>\n",
       "      <td>15833859</td>\n",
       "      <td>Gene</td>\n",
       "      <td>IFI27</td>\n",
       "      <td>NCBIGene:3429</td>\n",
       "      <td>physically_interacts_with</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Chemical API</td>\n",
       "      <td>18974147</td>\n",
       "      <td>Gene</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>name:IMATINIB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>759</th>\n",
       "      <td>CHRONIC MYELOGENOUS LEUKEMIA, BCR-ABL1 POSITIVE</td>\n",
       "      <td>Disease</td>\n",
       "      <td>related_to</td>\n",
       "      <td>Translator Text Mining Provider</td>\n",
       "      <td>CORD Disease API</td>\n",
       "      <td>None</td>\n",
       "      <td>Gene</td>\n",
       "      <td>KIT</td>\n",
       "      <td>NCBIGene:3815</td>\n",
       "      <td>related_to</td>\n",
       "      <td>Translator Text Mining Provider</td>\n",
       "      <td>CORD Chemical API</td>\n",
       "      <td>None</td>\n",
       "      <td>Gene</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>name:IMATINIB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>959</th>\n",
       "      <td>CHRONIC MYELOGENOUS LEUKEMIA, BCR-ABL1 POSITIVE</td>\n",
       "      <td>Disease</td>\n",
       "      <td>related_to</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Disease API</td>\n",
       "      <td>18577744,18984583,21154127,24091144,25965880</td>\n",
       "      <td>Gene</td>\n",
       "      <td>LYN</td>\n",
       "      <td>NCBIGene:4067</td>\n",
       "      <td>related_to</td>\n",
       "      <td>Translator Text Mining Provider</td>\n",
       "      <td>CORD Chemical API</td>\n",
       "      <td>None</td>\n",
       "      <td>Gene</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>name:IMATINIB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>935</th>\n",
       "      <td>CHRONIC MYELOGENOUS LEUKEMIA, BCR-ABL1 POSITIVE</td>\n",
       "      <td>Disease</td>\n",
       "      <td>related_to</td>\n",
       "      <td>Translator Text Mining Provider</td>\n",
       "      <td>CORD Disease API</td>\n",
       "      <td>None</td>\n",
       "      <td>Gene</td>\n",
       "      <td>SRC</td>\n",
       "      <td>NCBIGene:6714</td>\n",
       "      <td>physically_interacts_with</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Chemical API</td>\n",
       "      <td>27087366</td>\n",
       "      <td>Gene</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>name:IMATINIB</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                input input_type        pred1  \\\n",
       "1319  CHRONIC MYELOGENOUS LEUKEMIA, BCR-ABL1 POSITIVE    Disease  affected_by   \n",
       "563   CHRONIC MYELOGENOUS LEUKEMIA, BCR-ABL1 POSITIVE    Disease   related_to   \n",
       "1092  CHRONIC MYELOGENOUS LEUKEMIA, BCR-ABL1 POSITIVE    Disease    caused_by   \n",
       "1537  CHRONIC MYELOGENOUS LEUKEMIA, BCR-ABL1 POSITIVE    Disease   related_to   \n",
       "700   CHRONIC MYELOGENOUS LEUKEMIA, BCR-ABL1 POSITIVE    Disease   treated_by   \n",
       "911   CHRONIC MYELOGENOUS LEUKEMIA, BCR-ABL1 POSITIVE    Disease  affected_by   \n",
       "1264  CHRONIC MYELOGENOUS LEUKEMIA, BCR-ABL1 POSITIVE    Disease    caused_by   \n",
       "759   CHRONIC MYELOGENOUS LEUKEMIA, BCR-ABL1 POSITIVE    Disease   related_to   \n",
       "959   CHRONIC MYELOGENOUS LEUKEMIA, BCR-ABL1 POSITIVE    Disease   related_to   \n",
       "935   CHRONIC MYELOGENOUS LEUKEMIA, BCR-ABL1 POSITIVE    Disease   related_to   \n",
       "\n",
       "                         pred1_source           pred1_api  \\\n",
       "1319                           SEMMED  SEMMED Disease API   \n",
       "563                              None         BioLink API   \n",
       "1092                           SEMMED  SEMMED Disease API   \n",
       "1537                           SEMMED  SEMMED Disease API   \n",
       "700                            SEMMED  SEMMED Disease API   \n",
       "911                            SEMMED  SEMMED Disease API   \n",
       "1264                           SEMMED  SEMMED Disease API   \n",
       "759   Translator Text Mining Provider    CORD Disease API   \n",
       "959                            SEMMED  SEMMED Disease API   \n",
       "935   Translator Text Mining Provider    CORD Disease API   \n",
       "\n",
       "                                           pred1_pubmed node1_type node1_name  \\\n",
       "1319                                           27573699       Gene        MYC   \n",
       "563                                                None       Gene        BCR   \n",
       "1092  10092207,10224280,10490629,10676660,10833515,1...       Gene       MTTP   \n",
       "1537                                           10688886       Gene     SQSTM1   \n",
       "700   15717993,16205964,16596277,17694264,17891251,2...       Gene        KIT   \n",
       "911                                            22823957       Gene      ABCB1   \n",
       "1264                                           15833859       Gene      IFI27   \n",
       "759                                                None       Gene        KIT   \n",
       "959        18577744,18984583,21154127,24091144,25965880       Gene        LYN   \n",
       "935                                                None       Gene        SRC   \n",
       "\n",
       "           node1_id                      pred2  \\\n",
       "1319  NCBIGene:4609    positively_regulated_by   \n",
       "563    NCBIGene:613       positively_regulates   \n",
       "1092  NCBIGene:4547              coexists_with   \n",
       "1537  NCBIGene:8878                 related_to   \n",
       "700   NCBIGene:3815       positively_regulates   \n",
       "911   NCBIGene:5243  physically_interacts_with   \n",
       "1264  NCBIGene:3429  physically_interacts_with   \n",
       "759   NCBIGene:3815                 related_to   \n",
       "959   NCBIGene:4067                 related_to   \n",
       "935   NCBIGene:6714  physically_interacts_with   \n",
       "\n",
       "                         pred2_source            pred2_api  \\\n",
       "1319                           SEMMED  SEMMED Chemical API   \n",
       "563                            SEMMED  SEMMED Chemical API   \n",
       "1092                           SEMMED  SEMMED Chemical API   \n",
       "1537                              CTD              CTD API   \n",
       "700                            SEMMED  SEMMED Chemical API   \n",
       "911                            SEMMED  SEMMED Chemical API   \n",
       "1264                           SEMMED  SEMMED Chemical API   \n",
       "759   Translator Text Mining Provider    CORD Chemical API   \n",
       "959   Translator Text Mining Provider    CORD Chemical API   \n",
       "935                            SEMMED  SEMMED Chemical API   \n",
       "\n",
       "                                           pred2_pubmed output_type  \\\n",
       "1319                                           22842456        Gene   \n",
       "563                          20228846,22204395,22641215        Gene   \n",
       "1092       12669727,15283151,21505592,22713161,23434731        Gene   \n",
       "1537                                  22641616,30165626        Gene   \n",
       "700                          17139461,28765927,28878021        Gene   \n",
       "911   16157201,17495881,17690695,18669873,19225039,2...        Gene   \n",
       "1264                                           18974147        Gene   \n",
       "759                                                None        Gene   \n",
       "959                                                None        Gene   \n",
       "935                                            27087366        Gene   \n",
       "\n",
       "     output_name      output_id  \n",
       "1319    IMATINIB  name:IMATINIB  \n",
       "563     IMATINIB  name:IMATINIB  \n",
       "1092    IMATINIB  name:IMATINIB  \n",
       "1537    IMATINIB  name:IMATINIB  \n",
       "700     IMATINIB  name:IMATINIB  \n",
       "911     IMATINIB  name:IMATINIB  \n",
       "1264    IMATINIB  name:IMATINIB  \n",
       "759     IMATINIB  name:IMATINIB  \n",
       "959     IMATINIB  name:IMATINIB  \n",
       "935     IMATINIB  name:IMATINIB  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = fc.display_table_view()\n",
    "\n",
    "# because UMLS is not currently well-integrated in our ID-to-object translation system, removing UMLS-only entries here\n",
    "patternDel = \"^UMLS:C\\d+\"\n",
    "filter = df.node1_id.str.contains(patternDel)\n",
    "df = df[~filter]\n",
    "\n",
    "print(df.shape)\n",
    "df.sample(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Notice that results give the knowledge source (API) and that many also give the literature source (Pubmed ID). \n",
    "\n",
    "Next, let's look to see which genes are mentioned the most."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "BCR      210\n",
       "ABL1      78\n",
       "KIT       72\n",
       "MTTP      42\n",
       "AKT1      36\n",
       "TP53      25\n",
       "ABCG2     20\n",
       "ABCB1     18\n",
       "LYN       18\n",
       "PTEN      16\n",
       "Name: node1_name, dtype: int64"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.node1_name.value_counts().head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Note: as of 2020-08-13, the notebook output above matches the description below. As knowledge sources (APIs) and code are updated and added, users may obtain different results.*\n",
    "\n",
    "Not surprisingly, the top two genes that BioThings Explorer found that join imatinib to CML are *ABL1* and *BCR*, the two genes that are fused in the \"Philadelphia chromosome\", the genetic abnormality that underlies CML, *and* the validated target of imatinib.\n",
    "\n",
    "Let's examine some of the PubMed articles linking **CML to *ABL1*** and ***ABL1* to imatinib**."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 17 articles supporting the edge between CML and ABL1. Sampling of 10 of those:\n",
      "http://pubmed.gov/24662807\n",
      "http://pubmed.gov/26179066\n",
      "http://pubmed.gov/10498618\n",
      "http://pubmed.gov/10822991\n",
      "http://pubmed.gov/11368359\n",
      "http://pubmed.gov/11979553\n",
      "http://pubmed.gov/18082628\n",
      "http://pubmed.gov/18243808\n",
      "http://pubmed.gov/18406870\n",
      "http://pubmed.gov/20809971\n"
     ]
    }
   ],
   "source": [
    "# fetch all articles connecting 'chronic myelogenous leukemia' and 'ABL1'\n",
    "ABL1_str = 'ABL1'\n",
    "CML_name = cml['name']\n",
    "\n",
    "articles = []\n",
    "for info in fc.display_edge_info(CML_name, ABL1_str).values():\n",
    "    if 'pubmed' in info['info']:\n",
    "        articles += info['info']['pubmed']\n",
    "print(\"There are \"+str(len(articles))+\" articles supporting the edge between CML and ABL1. Sampling of 10 of those:\")\n",
    "x = [print(\"http://pubmed.gov/\" + str(x)) for x in articles[0:10] ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 29 articles supporting the edge between ABL1 and imatinib. Sampling of 10 of those:\n",
      "http://pubmed.gov/15799618\n",
      "http://pubmed.gov/15917650\n",
      "http://pubmed.gov/15949566\n",
      "http://pubmed.gov/16153117\n",
      "http://pubmed.gov/16205964\n",
      "http://pubmed.gov/15713800\n",
      "http://pubmed.gov/19166098\n",
      "http://pubmed.gov/26030291\n",
      "http://pubmed.gov/26618175\n",
      "http://pubmed.gov/21693657\n"
     ]
    }
   ],
   "source": [
    "# fetch all articles connecting 'ABL1' and 'Imatinib\n",
    "imatinib_name = imatinib['name']\n",
    "\n",
    "articles = []\n",
    "for info in fc.display_edge_info(ABL1_str, imatinib_name).values():\n",
    "    if 'pubmed' in info['info']:\n",
    "        articles += info['info']['pubmed']\n",
    "print(\"There are \"+str(len(articles))+\" articles supporting the edge between ABL1 and imatinib. Sampling of 10 of those:\")\n",
    "x = [print(\"http://pubmed.gov/\" + str(x)) for x in articles[0:10] ]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Comparing results between CML and GIST"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's perform another BioThings Explorer query, this time looking to EXPLAIN the relationship between imatinib and gastrointestinal stromal tumors (GIST), another disease treated by imatinib."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 4 BioThings objects returned for gastrointestinal stromal tumor:\n",
      "0: ['gastrointestinal stromal tumor', 'Disease']\n",
      "1: ['GASTROINTESTINAL STROMAL TUMOR, FAMILIAL', 'Disease']\n",
      "2: ['Gastric Gastrointestinal Stromal Tumor', 'Disease']\n",
      "3: ['colorectal gastrointestinal stromal tumor', 'Disease']\n"
     ]
    }
   ],
   "source": [
    "ht = Hint()\n",
    "gist_str = \"gastrointestinal stromal tumor\"\n",
    "\n",
    "# find all potential representations of CML\n",
    "gist_hint = ht.query(gist_str)\n",
    "\n",
    "hint_display(gist_str, gist_hint)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "While all results look valid, for now we'll pick the top Disease choice (indexed at 0) for GIST for our query. We can look at identifier mappings inside this BioThings object. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'MONDO': 'MONDO:0011719',\n",
       " 'DOID': 'DOID:9253',\n",
       " 'UMLS': 'C0238198',\n",
       " 'name': 'gastrointestinal stromal tumor',\n",
       " 'MESH': 'D046152',\n",
       " 'OMIM': '606764',\n",
       " 'ORPHANET': '44890',\n",
       " 'primary': {'identifier': 'MONDO',\n",
       "  'cls': 'Disease',\n",
       "  'value': 'MONDO:0011719'},\n",
       " 'display': 'MONDO(MONDO:0011719) DOID(DOID:9253) OMIM(606764) ORPHANET(44890) UMLS(C0238198) MESH(D046152) name(gastrointestinal stromal tumor)',\n",
       " 'type': 'Disease'}"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# select the correct representation of CML\n",
    "gist = gist_hint['Disease'][0]\n",
    "gist"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Notice that like before, we are specifying with the `intermediate_nodes` parameter that we are looking for paths joining GIST and imatinib with *one* intermediate node that is a \"Gene\" (BioThings type).  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "fc2 = FindConnection(input_obj = gist, output_obj = imatinib, \\\n",
    "                     intermediate_nodes = 'Gene')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==========\n",
      "========== QUERY PARAMETER SUMMARY ==========\n",
      "==========\n",
      "\n",
      "BTE will find paths that join 'gastrointestinal stromal tumor' and 'imatinib'. Paths will have 1 intermediate node.\n",
      "\n",
      "Intermediate node #1 will have these type constraints: Gene\n",
      "\n",
      "\n",
      "==========\n",
      "========== QUERY #1 -- fetch all Gene entities linked to 'gastrointestinal stromal tumor' ==========\n",
      "==========\n",
      "\n",
      "==== Step #1: Query path planning ====\n",
      "\n",
      "Because gastrointestinal stromal tumor is of type 'Disease', BTE will query our meta-KG for APIs that can take 'Disease' as input and 'Gene' as output\n",
      "\n",
      "BTE found 10 apis:\n",
      "\n",
      "API 1. pharos(1 API call)\n",
      "API 2. scigraph(1 API call)\n",
      "API 3. cord_disease(1 API call)\n",
      "API 4. semmed_disease(15 API calls)\n",
      "API 5. hetio(1 API call)\n",
      "API 6. DISEASES(1 API call)\n",
      "API 7. mgi_gene2phenotype(1 API call)\n",
      "API 8. mydisease(1 API call)\n",
      "API 9. biolink(1 API call)\n",
      "API 10. scibite(1 API call)\n",
      "\n",
      "\n",
      "==== Step #2: Query path execution ====\n",
      "NOTE: API requests are dispatched in parallel, so the list of APIs below is ordered by query time.\n",
      "\n",
      "API 7.1: https://mydisease.info/v1/query?fields=disgenet.genes_related_to_disease.gene_id (POST -d q=C0238198&scopes=mondo.xrefs.umls, disgenet.xrefs.umls)\n",
      "API 4.11: https://biothings.ncats.io/semmed/query?fields=negatively_regulates (POST -d q=C0238198&scopes=umls)\n",
      "API 4.7: https://biothings.ncats.io/semmed/query?fields=positively_regulated_by (POST -d q=C0238198&scopes=umls)\n",
      "API 4.5: https://biothings.ncats.io/semmed/query?fields=physically_interacts_with (POST -d q=C0238198&scopes=umls)\n",
      "API 4.3: https://biothings.ncats.io/semmed/query?fields=prevented_by (POST -d q=C0238198&scopes=umls)\n",
      "API 4.14: https://biothings.ncats.io/semmed/query?fields=caused_by (POST -d q=C0238198&scopes=umls)\n",
      "API 4.8: https://biothings.ncats.io/semmed/query?fields=derives_from (POST -d q=C0238198&scopes=umls)\n",
      "API 4.1: https://biothings.ncats.io/semmed/query?fields=affects (POST -d q=C0238198&scopes=umls)\n",
      "API 6.1: https://pending.biothings.io/DISEASES/query?fields=DISEASES.associatedWith (POST -d q=DOID:9253&scopes=DISEASES.doid)\n",
      "API 8.1: https://pending.biothings.io/mgigene2phenotype/query?fields=_id&size=300 (POST -d q=DOID:9253&scopes=mgi.associated_with_disease.doid)\n",
      "API 4.12: https://biothings.ncats.io/semmed/query?fields=related_to (POST -d q=C0238198&scopes=umls)\n",
      "API 4.13: https://biothings.ncats.io/semmed/query?fields=disrupts (POST -d q=C0238198&scopes=umls)\n",
      "API 4.4: https://biothings.ncats.io/semmed/query?fields=positively_regulates (POST -d q=C0238198&scopes=umls)\n",
      "API 4.15: https://biothings.ncats.io/semmed/query?fields=negatively_regulated_by (POST -d q=C0238198&scopes=umls)\n",
      "API 3.1: https://biothings.ncats.io/cord_disease/query?fields=associated_with (POST -d q=DOID:9253&scopes=doid)\n",
      "API 4.2: https://biothings.ncats.io/semmed/query?fields=coexists_with (POST -d q=C0238198&scopes=umls)\n",
      "API 4.10: https://biothings.ncats.io/semmed/query?fields=affected_by (POST -d q=C0238198&scopes=umls)\n",
      "API 4.9: https://biothings.ncats.io/semmed/query?fields=disrupted_by (POST -d q=C0238198&scopes=umls)\n",
      "API 4.6: https://biothings.ncats.io/semmed/query?fields=treated_by (POST -d q=C0238198&scopes=umls)\n",
      "API 9.1: https://api.monarchinitiative.org/api/bioentity/disease/MONDO:0011719/genes?rows=200\n",
      "API 2.1: https://automat.renci.org/cord19_scigraph_v2/disease/gene/MONDO:0011719\n",
      "API 2.1 scigraph failed\n",
      "API 10.1: https://automat.renci.org/cord19_scibite_v2/disease/gene/MONDO:0011719\n",
      "API 10.1 scibite failed\n",
      "API 5.1: https://automat.renci.org/hetio/disease/gene/MONDO:0011719\n",
      "API 1.1: https://automat.renci.org/pharos/disease/gene/MONDO:0011719\n",
      "\n",
      "\n",
      "==== Step #3: Output normalization ====\n",
      "\n",
      "API 8.1 mgi_gene2phenotype: 1 hits\n",
      "API 4.1 semmed_disease: No hits\n",
      "API 4.2 semmed_disease: 1 hits\n",
      "API 6.1 DISEASES: 38 hits\n",
      "API 4.3 semmed_disease: 1 hits\n",
      "API 4.4 semmed_disease: No hits\n",
      "API 4.5 semmed_disease: 1 hits\n",
      "API 4.6 semmed_disease: 53 hits\n",
      "API 4.7 semmed_disease: No hits\n",
      "API 1.1 pharos: 30 hits\n",
      "API 4.8 semmed_disease: No hits\n",
      "API 4.9 semmed_disease: 4 hits\n",
      "API 3.1 cord_disease: 64 hits\n",
      "API 4.10 semmed_disease: 18 hits\n",
      "API 5.1 hetio: No hits\n",
      "API 4.11 semmed_disease: No hits\n",
      "API 4.12 semmed_disease: 250 hits\n",
      "API 7.1 mydisease: 51 hits\n",
      "API 4.13 semmed_disease: No hits\n",
      "API 4.14 semmed_disease: 17 hits\n",
      "API 9.1 biolink: 5 hits\n",
      "API 4.15 semmed_disease: No hits\n",
      "\n",
      "After id-to-object translation, BTE retrieved 385 unique objects.\n",
      "\n",
      "\n",
      "==========\n",
      "========== QUERY #2 -- fetch all Gene entities linked to 'imatinib' ==========\n",
      "==========\n",
      "\n",
      "==== Step #1: Query path planning ====\n",
      "\n",
      "Because imatinib is of type 'ChemicalSubstance', BTE will query our meta-KG for APIs that can take 'ChemicalSubstance' as input and 'Gene' as output\n",
      "\n",
      "BTE found 10 apis:\n",
      "\n",
      "API 1. cord_chemical(1 API call)\n",
      "API 2. pharos(1 API call)\n",
      "API 3. scigraph(1 API call)\n",
      "API 4. mychem(3 API calls)\n",
      "API 5. chembio(1 API call)\n",
      "API 6. ctd(2 API calls)\n",
      "API 7. semmed_chemical(13 API calls)\n",
      "API 8. dgidb(2 API calls)\n",
      "API 9. hmdb(1 API call)\n",
      "API 10. scibite(1 API call)\n",
      "\n",
      "\n",
      "==== Step #2: Query path execution ====\n",
      "NOTE: API requests are dispatched in parallel, so the list of APIs below is ordered by query time.\n",
      "\n",
      "API 8.1: http://dgidb.genome.wustl.edu/api/v2/interactions.json?drugs=CHEMBL1642\n",
      "API 7.2: https://biothings.ncats.io/semmedchemical/query?fields=positively_regulates (POST -d q=C0935989&scopes=umls)\n",
      "API 7.12: https://biothings.ncats.io/semmedchemical/query?fields=affected_by (POST -d q=C0935989&scopes=umls)\n",
      "API 6.2: https://mychem.info/v1/query?fields=drugcentral.bioactivity (POST -d q=CHEMBL1642,CHEMBL941&scopes=chembl.molecule_chembl_id)\n",
      "API 7.1: https://biothings.ncats.io/semmedchemical/query?fields=physically_interacts_with (POST -d q=C0935989&scopes=umls)\n",
      "API 7.8: https://biothings.ncats.io/semmedchemical/query?fields=produces (POST -d q=C0935989&scopes=umls)\n",
      "API 7.5: https://biothings.ncats.io/semmedchemical/query?fields=related_to (POST -d q=C0935989&scopes=umls)\n",
      "API 7.3: https://biothings.ncats.io/semmedchemical/query?fields=negatively_regulated_by (POST -d q=C0935989&scopes=umls)\n",
      "API 7.6: https://biothings.ncats.io/semmedchemical/query?fields=disrupts (POST -d q=C0935989&scopes=umls)\n",
      "API 7.11: https://biothings.ncats.io/semmedchemical/query?fields=affects (POST -d q=C0935989&scopes=umls)\n",
      "API 1.1: https://biothings.ncats.io/cord_chemical/query?fields=associated_with (POST -d q=CHEBI:45783&scopes=chebi)\n",
      "API 7.7: https://biothings.ncats.io/semmedchemical/query?fields=coexists_with (POST -d q=C0935989&scopes=umls)\n",
      "API 7.9: https://biothings.ncats.io/semmedchemical/query?fields=produced_by (POST -d q=C0935989&scopes=umls)\n",
      "API 7.4: https://biothings.ncats.io/semmedchemical/query?fields=disrupted_by (POST -d q=C0935989&scopes=umls)\n",
      "API 7.10: https://biothings.ncats.io/semmedchemical/query?fields=negatively_regulates (POST -d q=C0935989&scopes=umls)\n",
      "API 7.13: https://biothings.ncats.io/semmedchemical/query?fields=positively_regulated_by (POST -d q=C0935989&scopes=umls)\n",
      "API 6.1: https://mychem.info/v1/query?fields=drugbank.targets (POST -d q=DB00619&scopes=drugbank.id)\n",
      "API 6.3: https://mychem.info/v1/query?fields=drugbank.enzymes (POST -d q=DB00619&scopes=drugbank.id)\n",
      "API 4.2: http://ctdbase.org/tools/batchQuery.go?inputType=chem&report=genes_curated&format=json&inputTerms=C097613\n",
      "API 4.1: http://ctdbase.org/tools/batchQuery.go?inputType=chem&report=genes_curated&format=json&inputTerms=D000068877\n",
      "API 8.2: http://dgidb.genome.wustl.edu/api/v2/interactions.json?drugs=CHEMBL941\n",
      "API 10.1: https://automat.renci.org/cord19_scibite_v2/chemical_substance/gene/CHEBI:45783\n",
      "API 10.1 scibite failed\n",
      "API 3.1: https://automat.renci.org/cord19_scigraph_v2/chemical_substance/gene/CHEBI:45783\n",
      "API 3.1 scigraph failed\n",
      "API 5.1: https://automat.renci.org/chembio/chemical_substance/gene/CHEBI:45783\n",
      "API 9.1: https://automat.renci.org/hmdb/chemical_substance/gene/CHEBI:45783\n",
      "API 2.1: https://automat.renci.org/pharos/chemical_substance/gene/CHEBI:45783\n",
      "\n",
      "\n",
      "==== Step #3: Output normalization ====\n",
      "\n",
      "API 7.1 semmed_chemical: 226 hits\n",
      "API 5.1 chembio: No hits\n",
      "API 6.1 mychem: 9 hits\n",
      "API 6.2 mychem: 80 hits\n",
      "API 6.3 mychem: 9 hits\n",
      "API 2.1 pharos: 6 hits\n",
      "API 7.2 semmed_chemical: 96 hits\n",
      "API 7.3 semmed_chemical: 36 hits\n",
      "API 7.4 semmed_chemical: No hits\n",
      "API 7.5 semmed_chemical: No hits\n",
      "API 8.1 dgidb: 5 hits\n",
      "API 8.2 dgidb: 34 hits\n",
      "API 9.1 hmdb: No hits\n",
      "API 7.6 semmed_chemical: No hits\n",
      "API 7.7 semmed_chemical: 119 hits\n",
      "API 7.8 semmed_chemical: No hits\n",
      "API 7.9 semmed_chemical: 11 hits\n",
      "API 1.1 cord_chemical: 172 hits\n",
      "API 4.1 ctd: 199 hits\n",
      "API 4.2 ctd: No hits\n",
      "API 7.10 semmed_chemical: 166 hits\n",
      "API 7.11 semmed_chemical: No hits\n",
      "API 7.12 semmed_chemical: No hits\n",
      "API 7.13 semmed_chemical: 39 hits\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "After id-to-object translation, BTE retrieved 728 unique objects.\n",
      "\n",
      "==========\n",
      "========== Final assembly of results ==========\n",
      "==========\n",
      "\n",
      "\n",
      "BTE found 130 unique intermediate nodes connecting 'gastrointestinal stromal tumor' and 'imatinib'\n"
     ]
    }
   ],
   "source": [
    "fc2.connect(verbose = True) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(772, 16)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>input</th>\n",
       "      <th>input_type</th>\n",
       "      <th>pred1</th>\n",
       "      <th>pred1_source</th>\n",
       "      <th>pred1_api</th>\n",
       "      <th>pred1_pubmed</th>\n",
       "      <th>node1_type</th>\n",
       "      <th>node1_name</th>\n",
       "      <th>node1_id</th>\n",
       "      <th>pred2</th>\n",
       "      <th>pred2_source</th>\n",
       "      <th>pred2_api</th>\n",
       "      <th>pred2_pubmed</th>\n",
       "      <th>output_type</th>\n",
       "      <th>output_name</th>\n",
       "      <th>output_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>694</th>\n",
       "      <td>GASTROINTESTINAL STROMAL TUMOR</td>\n",
       "      <td>Disease</td>\n",
       "      <td>treated_by</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Disease API</td>\n",
       "      <td>24399735</td>\n",
       "      <td>Gene</td>\n",
       "      <td>CYP3A4</td>\n",
       "      <td>NCBIGene:1576</td>\n",
       "      <td>related_to</td>\n",
       "      <td>Translator Text Mining Provider</td>\n",
       "      <td>CORD Chemical API</td>\n",
       "      <td>None</td>\n",
       "      <td>Gene</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>name:IMATINIB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>707</th>\n",
       "      <td>GASTROINTESTINAL STROMAL TUMOR</td>\n",
       "      <td>Disease</td>\n",
       "      <td>related_to</td>\n",
       "      <td>Translator Text Mining Provider</td>\n",
       "      <td>CORD Disease API</td>\n",
       "      <td>None</td>\n",
       "      <td>Gene</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>NCBIGene:1956</td>\n",
       "      <td>physically_interacts_with</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Chemical API</td>\n",
       "      <td>15887238,16476837</td>\n",
       "      <td>Gene</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>name:IMATINIB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>675</th>\n",
       "      <td>GASTROINTESTINAL STROMAL TUMOR</td>\n",
       "      <td>Disease</td>\n",
       "      <td>related_to</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Disease API</td>\n",
       "      <td>18265649,18648736,25525851,28698433</td>\n",
       "      <td>Gene</td>\n",
       "      <td>PDGFRB</td>\n",
       "      <td>NCBIGene:5159</td>\n",
       "      <td>physically_interacts_with</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Chemical API</td>\n",
       "      <td>28028030</td>\n",
       "      <td>Gene</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>name:IMATINIB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>919</th>\n",
       "      <td>GASTROINTESTINAL STROMAL TUMOR</td>\n",
       "      <td>Disease</td>\n",
       "      <td>related_to</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Disease API</td>\n",
       "      <td>20028860</td>\n",
       "      <td>Gene</td>\n",
       "      <td>H2AX</td>\n",
       "      <td>NCBIGene:3014</td>\n",
       "      <td>physically_interacts_with</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Chemical API</td>\n",
       "      <td>22388075</td>\n",
       "      <td>Gene</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>name:IMATINIB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>990</th>\n",
       "      <td>GASTROINTESTINAL STROMAL TUMOR</td>\n",
       "      <td>Disease</td>\n",
       "      <td>related_to</td>\n",
       "      <td>Translator Text Mining Provider</td>\n",
       "      <td>CORD Disease API</td>\n",
       "      <td>None</td>\n",
       "      <td>Gene</td>\n",
       "      <td>BCL2</td>\n",
       "      <td>NCBIGene:596</td>\n",
       "      <td>related_to</td>\n",
       "      <td>CTD</td>\n",
       "      <td>CTD API</td>\n",
       "      <td>23994742</td>\n",
       "      <td>Gene</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>name:IMATINIB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>975</th>\n",
       "      <td>GASTROINTESTINAL STROMAL TUMOR</td>\n",
       "      <td>Disease</td>\n",
       "      <td>related_to</td>\n",
       "      <td>DISEASE</td>\n",
       "      <td>DISEASES API</td>\n",
       "      <td>None</td>\n",
       "      <td>Gene</td>\n",
       "      <td>SPINK5</td>\n",
       "      <td>NCBIGene:11005</td>\n",
       "      <td>related_to</td>\n",
       "      <td>Translator Text Mining Provider</td>\n",
       "      <td>CORD Chemical API</td>\n",
       "      <td>None</td>\n",
       "      <td>Gene</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>name:IMATINIB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>431</th>\n",
       "      <td>GASTROINTESTINAL STROMAL TUMOR</td>\n",
       "      <td>Disease</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Disease API</td>\n",
       "      <td>10779223,11706520,12938260,27771813</td>\n",
       "      <td>Gene</td>\n",
       "      <td>KIT</td>\n",
       "      <td>NCBIGene:3815</td>\n",
       "      <td>negatively_regulates</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Chemical API</td>\n",
       "      <td>12938280,14531349,15139047,15966213,16614006,1...</td>\n",
       "      <td>Gene</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>name:IMATINIB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>730</th>\n",
       "      <td>GASTROINTESTINAL STROMAL TUMOR</td>\n",
       "      <td>Disease</td>\n",
       "      <td>related_to</td>\n",
       "      <td>DISEASE</td>\n",
       "      <td>DISEASES API</td>\n",
       "      <td>None</td>\n",
       "      <td>Gene</td>\n",
       "      <td>BRAF</td>\n",
       "      <td>NCBIGene:673</td>\n",
       "      <td>physically_interacts_with</td>\n",
       "      <td>drugcentral</td>\n",
       "      <td>MyChem.info API</td>\n",
       "      <td>None</td>\n",
       "      <td>Gene</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>name:IMATINIB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200</th>\n",
       "      <td>GASTROINTESTINAL STROMAL TUMOR</td>\n",
       "      <td>Disease</td>\n",
       "      <td>related_to</td>\n",
       "      <td>Translator Text Mining Provider</td>\n",
       "      <td>CORD Disease API</td>\n",
       "      <td>None</td>\n",
       "      <td>Gene</td>\n",
       "      <td>BCR</td>\n",
       "      <td>NCBIGene:613</td>\n",
       "      <td>related_to</td>\n",
       "      <td>CTD</td>\n",
       "      <td>CTD API</td>\n",
       "      <td>20222756</td>\n",
       "      <td>Gene</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>name:IMATINIB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>965</th>\n",
       "      <td>GASTROINTESTINAL STROMAL TUMOR</td>\n",
       "      <td>Disease</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Disease API</td>\n",
       "      <td>28862263</td>\n",
       "      <td>Gene</td>\n",
       "      <td>NF1</td>\n",
       "      <td>NCBIGene:4763</td>\n",
       "      <td>related_to</td>\n",
       "      <td>Translator Text Mining Provider</td>\n",
       "      <td>CORD Chemical API</td>\n",
       "      <td>None</td>\n",
       "      <td>Gene</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>name:IMATINIB</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              input input_type       pred1  \\\n",
       "694  GASTROINTESTINAL STROMAL TUMOR    Disease  treated_by   \n",
       "707  GASTROINTESTINAL STROMAL TUMOR    Disease  related_to   \n",
       "675  GASTROINTESTINAL STROMAL TUMOR    Disease  related_to   \n",
       "919  GASTROINTESTINAL STROMAL TUMOR    Disease  related_to   \n",
       "990  GASTROINTESTINAL STROMAL TUMOR    Disease  related_to   \n",
       "975  GASTROINTESTINAL STROMAL TUMOR    Disease  related_to   \n",
       "431  GASTROINTESTINAL STROMAL TUMOR    Disease   caused_by   \n",
       "730  GASTROINTESTINAL STROMAL TUMOR    Disease  related_to   \n",
       "200  GASTROINTESTINAL STROMAL TUMOR    Disease  related_to   \n",
       "965  GASTROINTESTINAL STROMAL TUMOR    Disease   caused_by   \n",
       "\n",
       "                        pred1_source           pred1_api  \\\n",
       "694                           SEMMED  SEMMED Disease API   \n",
       "707  Translator Text Mining Provider    CORD Disease API   \n",
       "675                           SEMMED  SEMMED Disease API   \n",
       "919                           SEMMED  SEMMED Disease API   \n",
       "990  Translator Text Mining Provider    CORD Disease API   \n",
       "975                          DISEASE        DISEASES API   \n",
       "431                           SEMMED  SEMMED Disease API   \n",
       "730                          DISEASE        DISEASES API   \n",
       "200  Translator Text Mining Provider    CORD Disease API   \n",
       "965                           SEMMED  SEMMED Disease API   \n",
       "\n",
       "                            pred1_pubmed node1_type node1_name  \\\n",
       "694                             24399735       Gene     CYP3A4   \n",
       "707                                 None       Gene       EGFR   \n",
       "675  18265649,18648736,25525851,28698433       Gene     PDGFRB   \n",
       "919                             20028860       Gene       H2AX   \n",
       "990                                 None       Gene       BCL2   \n",
       "975                                 None       Gene     SPINK5   \n",
       "431  10779223,11706520,12938260,27771813       Gene        KIT   \n",
       "730                                 None       Gene       BRAF   \n",
       "200                                 None       Gene        BCR   \n",
       "965                             28862263       Gene        NF1   \n",
       "\n",
       "           node1_id                      pred2  \\\n",
       "694   NCBIGene:1576                 related_to   \n",
       "707   NCBIGene:1956  physically_interacts_with   \n",
       "675   NCBIGene:5159  physically_interacts_with   \n",
       "919   NCBIGene:3014  physically_interacts_with   \n",
       "990    NCBIGene:596                 related_to   \n",
       "975  NCBIGene:11005                 related_to   \n",
       "431   NCBIGene:3815       negatively_regulates   \n",
       "730    NCBIGene:673  physically_interacts_with   \n",
       "200    NCBIGene:613                 related_to   \n",
       "965   NCBIGene:4763                 related_to   \n",
       "\n",
       "                        pred2_source            pred2_api  \\\n",
       "694  Translator Text Mining Provider    CORD Chemical API   \n",
       "707                           SEMMED  SEMMED Chemical API   \n",
       "675                           SEMMED  SEMMED Chemical API   \n",
       "919                           SEMMED  SEMMED Chemical API   \n",
       "990                              CTD              CTD API   \n",
       "975  Translator Text Mining Provider    CORD Chemical API   \n",
       "431                           SEMMED  SEMMED Chemical API   \n",
       "730                      drugcentral      MyChem.info API   \n",
       "200                              CTD              CTD API   \n",
       "965  Translator Text Mining Provider    CORD Chemical API   \n",
       "\n",
       "                                          pred2_pubmed output_type  \\\n",
       "694                                               None        Gene   \n",
       "707                                  15887238,16476837        Gene   \n",
       "675                                           28028030        Gene   \n",
       "919                                           22388075        Gene   \n",
       "990                                           23994742        Gene   \n",
       "975                                               None        Gene   \n",
       "431  12938280,14531349,15139047,15966213,16614006,1...        Gene   \n",
       "730                                               None        Gene   \n",
       "200                                           20222756        Gene   \n",
       "965                                               None        Gene   \n",
       "\n",
       "    output_name      output_id  \n",
       "694    IMATINIB  name:IMATINIB  \n",
       "707    IMATINIB  name:IMATINIB  \n",
       "675    IMATINIB  name:IMATINIB  \n",
       "919    IMATINIB  name:IMATINIB  \n",
       "990    IMATINIB  name:IMATINIB  \n",
       "975    IMATINIB  name:IMATINIB  \n",
       "431    IMATINIB  name:IMATINIB  \n",
       "730    IMATINIB  name:IMATINIB  \n",
       "200    IMATINIB  name:IMATINIB  \n",
       "965    IMATINIB  name:IMATINIB  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2 = fc2.display_table_view()\n",
    "\n",
    "# because UMLS is not currently well-integrated in our ID-to-object translation system, removing UMLS-only entries here\n",
    "patternDel = \"^UMLS:C\\d+\"\n",
    "filter = df2.node1_id.str.contains(patternDel)\n",
    "df2 = df2[~filter]\n",
    "\n",
    "print(df2.shape)\n",
    "df2.sample(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "KIT       306\n",
       "PDGFRA     88\n",
       "BCR        45\n",
       "VEGFA      24\n",
       "BRAF       20\n",
       "TP53       20\n",
       "MTTP       14\n",
       "ABL1       13\n",
       "CD34       12\n",
       "FIP1L1     12\n",
       "Name: node1_name, dtype: int64"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2.node1_name.value_counts().head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Note: as of 2020-08-13, the notebook output above matches the description below. As knowledge sources (APIs) and code are updated and added, users may obtain different results.*\n",
    "\n",
    "Here, the top two genes that BioThings Explorer found that join imatinib to GIST are *PDGFRA* and *KIT*, the most commonly mutated genes found in GIST *and* validated targets of imatinib.\n",
    "\n",
    "While several of the listed genes would be considered positive controls, others on the list could be viewed as **testable hypotheses and discovery opportunities** to be evaluated by domain experts."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Conclusions and caveats"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This notebook demonstrated the use of BioThings Explorer in EXPLAIN mode to investigate the relationship between imatinib and two diseases that it treats -- chronic myelogenous leukemia (CML) and gastrointestinal stromal tumors (GIST).  In each case, BioThings Explorer autonomously queried a **distributed knowledge graph of biomedical APIs** to find the most common genes, and in each case the relevant targets were retrieved.\n",
    "\n",
    "There are still many areas for improvement (and some areas in which BioThings Explorer is still buggy).  And of course, BioThings Explorer is dependent on the accessibility of the APIs that comprise the distributed knowledge graph.  Nevertheless, we encourage users to try other variants of the EXPLAIN queries demonstrated in this notebook."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## More query information for users"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Note: as of 2020-08-13, the information below is valid.*\n",
    "\n",
    "The input into the Hint query should be one string with **at least one whole word.** \n",
    "\n",
    "Currently, the user must set the input_obj parameter of FindConnection() as **one specific BioThings object** (ex: one chemical, phenotype, disease, GO term, etc.). To do an EXPLAIN query, the output_obj parameter of FindConnection() must be **another specific BioThings object**. \n",
    "\n",
    "**For an EXPLAIN query, BTE will only accept one intermediate node.** The intermediate node cannot be a specific BioThings object. They must be BioThings type(s), None, or an empty list ([]). \n",
    "\n",
    "The intermediate node parameter can be a single entity type (string, example: 'BiologicalEntity'), a list of acceptable entity types (list of strings, example: ['Gene', 'AnatomicalEntity']), None, or an empty list. **None or an empty list will tell BTE to look for direct connections (no intermediate nodes).** If the intermediate node parameter is a list, BTE will look for knowledge sources that can handle at least one of the types listed.   \n",
    "\n",
    "BTE will query from input_obj -> intermediate and **from output_obj -> intermediate** and then join results based on shared intermediates. The results should therefore be read as **input_obj -> intermediate <- output_obj. This directionality in results should be kept in mind when interpreting them.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.8"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
