{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Introduction\n",
    "\n",
    "This notebook demonstrates basic usage of BioThings Explorer, an engine for autonomously querying a distributed knowledge graph. BioThings Explorer can answer two classes of queries -- \"EXPLAIN\" and \"PREDICT\".  EXPLAIN queries are described in [EXPLAIN_demo.ipynb](EXPLAIN_demo.ipynb). Here, we describe PREDICT queries and how to use BioThings Explorer to execute them.  A more detailed overview of the BioThings Explorer systems is provided in [these slides](https://docs.google.com/presentation/d/1QWQqqQhPD_pzKryh6Wijm4YQswv8pAjleVORCPyJyDE/edit?usp=sharing).\n",
    "\n",
    "PREDICT queries are designed to **predict plausible relationships between one entity and an entity class**.  For example, in this notebook, we explore the question:\n",
    "\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;*\"What drugs might be used to treat hyperphenylalaninemia?\"*\n",
    "\n",
    "**To experiment with an executable version of this notebook, [load it in Google Colaboratory](https://colab.research.google.com/github/biothings/biothings_explorer/blob/master/jupyter%20notebooks/PREDICT_demo.ipynb).**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 0: Load BioThings Explorer modules"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First, install the `biothings_explorer` and `biothings_schema` packages, as described in this [README](https://github.com/biothings/biothings_explorer/blob/master/jupyter%20notebooks/README.md#prerequisite).  This only needs to be done once (but including it here for compability with [colab](https://colab.research.google.com/))."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: biothings_explorer from git+https://github.com/biothings/biothings_explorer#egg=biothings_explorer in ./ENV/lib/python3.7/site-packages (0.0.1)\n",
      "Requirement already satisfied: jsonpath-rw>=1.4.0 in ./ENV/lib/python3.7/site-packages (from biothings_explorer) (1.4.0)\n",
      "Requirement already satisfied: networkx==2.3 in ./ENV/lib/python3.7/site-packages (from biothings_explorer) (2.3)\n",
      "Requirement already satisfied: requests>=2.21.0 in ./ENV/lib/python3.7/site-packages (from biothings_explorer) (2.22.0)\n",
      "Requirement already satisfied: aiohttp in ./ENV/lib/python3.7/site-packages (from biothings_explorer) (3.5.4)\n",
      "Requirement already satisfied: pandas in ./ENV/lib/python3.7/site-packages (from biothings_explorer) (0.25.1)\n",
      "Requirement already satisfied: jsonschema>=3.0.1 in ./ENV/lib/python3.7/site-packages (from biothings_explorer) (3.0.1)\n",
      "Requirement already satisfied: biothings-schema@ git+https://github.com/biothings/biothings_schema.py from git+https://github.com/biothings/biothings_schema.py in ./ENV/lib/python3.7/site-packages (from biothings_explorer) (0.0.1)\n",
      "Requirement already satisfied: tornado==4.5.3 in ./ENV/lib/python3.7/site-packages (from biothings_explorer) (4.5.3)\n",
      "Requirement already satisfied: notebook==5.7.5 in ./ENV/lib/python3.7/site-packages (from biothings_explorer) (5.7.5)\n",
      "Requirement already satisfied: jupyter in ./ENV/lib/python3.7/site-packages (from biothings_explorer) (1.0.0)\n",
      "Requirement already satisfied: graphviz>=0.11.1 in ./ENV/lib/python3.7/site-packages (from biothings_explorer) (0.11.1)\n",
      "Requirement already satisfied: six in ./ENV/lib/python3.7/site-packages (from jsonpath-rw>=1.4.0->biothings_explorer) (1.12.0)\n",
      "Requirement already satisfied: ply in ./ENV/lib/python3.7/site-packages (from jsonpath-rw>=1.4.0->biothings_explorer) (3.11)\n",
      "Requirement already satisfied: decorator in ./ENV/lib/python3.7/site-packages (from jsonpath-rw>=1.4.0->biothings_explorer) (4.4.0)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in ./ENV/lib/python3.7/site-packages (from requests>=2.21.0->biothings_explorer) (2019.3.9)\n",
      "Requirement already satisfied: chardet<3.1.0,>=3.0.2 in ./ENV/lib/python3.7/site-packages (from requests>=2.21.0->biothings_explorer) (3.0.4)\n",
      "Requirement already satisfied: urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1 in ./ENV/lib/python3.7/site-packages (from requests>=2.21.0->biothings_explorer) (1.25.3)\n",
      "Requirement already satisfied: idna<2.9,>=2.5 in ./ENV/lib/python3.7/site-packages (from requests>=2.21.0->biothings_explorer) (2.8)\n",
      "Requirement already satisfied: multidict<5.0,>=4.0 in ./ENV/lib/python3.7/site-packages (from aiohttp->biothings_explorer) (4.5.2)\n",
      "Requirement already satisfied: async-timeout<4.0,>=3.0 in ./ENV/lib/python3.7/site-packages (from aiohttp->biothings_explorer) (3.0.1)\n",
      "Requirement already satisfied: attrs>=17.3.0 in ./ENV/lib/python3.7/site-packages (from aiohttp->biothings_explorer) (19.1.0)\n",
      "Requirement already satisfied: yarl<2.0,>=1.0 in ./ENV/lib/python3.7/site-packages (from aiohttp->biothings_explorer) (1.3.0)\n",
      "Requirement already satisfied: pytz>=2017.2 in ./ENV/lib/python3.7/site-packages (from pandas->biothings_explorer) (2019.2)\n",
      "Requirement already satisfied: python-dateutil>=2.6.1 in ./ENV/lib/python3.7/site-packages (from pandas->biothings_explorer) (2.8.0)\n",
      "Requirement already satisfied: numpy>=1.13.3 in ./ENV/lib/python3.7/site-packages (from pandas->biothings_explorer) (1.17.0)\n",
      "Requirement already satisfied: setuptools in ./ENV/lib/python3.7/site-packages (from jsonschema>=3.0.1->biothings_explorer) (41.0.1)\n",
      "Requirement already satisfied: pyrsistent>=0.14.0 in ./ENV/lib/python3.7/site-packages (from jsonschema>=3.0.1->biothings_explorer) (0.15.2)\n",
      "Requirement already satisfied: PyYAML>=5.1 in ./ENV/lib/python3.7/site-packages (from biothings-schema@ git+https://github.com/biothings/biothings_schema.py->biothings_explorer) (5.1.1)\n",
      "Requirement already satisfied: strict-rfc3339 in ./ENV/lib/python3.7/site-packages (from biothings-schema@ git+https://github.com/biothings/biothings_schema.py->biothings_explorer) (0.7)\n",
      "Requirement already satisfied: jupyter-core>=4.4.0 in ./ENV/lib/python3.7/site-packages (from notebook==5.7.5->biothings_explorer) (4.4.0)\n",
      "Requirement already satisfied: terminado>=0.8.1 in ./ENV/lib/python3.7/site-packages (from notebook==5.7.5->biothings_explorer) (0.8.2)\n",
      "Requirement already satisfied: prometheus-client in ./ENV/lib/python3.7/site-packages (from notebook==5.7.5->biothings_explorer) (0.7.0)\n",
      "Requirement already satisfied: pyzmq>=17 in ./ENV/lib/python3.7/site-packages (from notebook==5.7.5->biothings_explorer) (18.0.1)\n",
      "Requirement already satisfied: traitlets>=4.2.1 in ./ENV/lib/python3.7/site-packages (from notebook==5.7.5->biothings_explorer) (4.3.2)\n",
      "Requirement already satisfied: ipython-genutils in ./ENV/lib/python3.7/site-packages (from notebook==5.7.5->biothings_explorer) (0.2.0)\n",
      "Requirement already satisfied: nbformat in ./ENV/lib/python3.7/site-packages (from notebook==5.7.5->biothings_explorer) (4.4.0)\n",
      "Requirement already satisfied: nbconvert in ./ENV/lib/python3.7/site-packages (from notebook==5.7.5->biothings_explorer) (5.5.0)\n",
      "Requirement already satisfied: Send2Trash in ./ENV/lib/python3.7/site-packages (from notebook==5.7.5->biothings_explorer) (1.5.0)\n",
      "Requirement already satisfied: ipykernel in ./ENV/lib/python3.7/site-packages (from notebook==5.7.5->biothings_explorer) (5.1.1)\n",
      "Requirement already satisfied: jinja2 in ./ENV/lib/python3.7/site-packages (from notebook==5.7.5->biothings_explorer) (2.10.1)\n",
      "Requirement already satisfied: jupyter-client>=5.2.0 in ./ENV/lib/python3.7/site-packages (from notebook==5.7.5->biothings_explorer) (5.2.4)\n",
      "Requirement already satisfied: qtconsole in ./ENV/lib/python3.7/site-packages (from jupyter->biothings_explorer) (4.5.1)\n",
      "Requirement already satisfied: ipywidgets in ./ENV/lib/python3.7/site-packages (from jupyter->biothings_explorer) (7.4.2)\n",
      "Requirement already satisfied: jupyter-console in ./ENV/lib/python3.7/site-packages (from jupyter->biothings_explorer) (6.0.0)\n",
      "Requirement already satisfied: ptyprocess; os_name != \"nt\" in ./ENV/lib/python3.7/site-packages (from terminado>=0.8.1->notebook==5.7.5->biothings_explorer) (0.6.0)\n",
      "Requirement already satisfied: testpath in ./ENV/lib/python3.7/site-packages (from nbconvert->notebook==5.7.5->biothings_explorer) (0.4.2)\n",
      "Requirement already satisfied: defusedxml in ./ENV/lib/python3.7/site-packages (from nbconvert->notebook==5.7.5->biothings_explorer) (0.6.0)\n",
      "Requirement already satisfied: mistune>=0.8.1 in ./ENV/lib/python3.7/site-packages (from nbconvert->notebook==5.7.5->biothings_explorer) (0.8.4)\n",
      "Requirement already satisfied: bleach in ./ENV/lib/python3.7/site-packages (from nbconvert->notebook==5.7.5->biothings_explorer) (3.1.0)\n",
      "Requirement already satisfied: pandocfilters>=1.4.1 in ./ENV/lib/python3.7/site-packages (from nbconvert->notebook==5.7.5->biothings_explorer) (1.4.2)\n",
      "Requirement already satisfied: entrypoints>=0.2.2 in ./ENV/lib/python3.7/site-packages (from nbconvert->notebook==5.7.5->biothings_explorer) (0.3)\n",
      "Requirement already satisfied: pygments in ./ENV/lib/python3.7/site-packages (from nbconvert->notebook==5.7.5->biothings_explorer) (2.4.2)\n",
      "Requirement already satisfied: ipython>=5.0.0 in ./ENV/lib/python3.7/site-packages (from ipykernel->notebook==5.7.5->biothings_explorer) (7.5.0)\n",
      "Requirement already satisfied: MarkupSafe>=0.23 in ./ENV/lib/python3.7/site-packages (from jinja2->notebook==5.7.5->biothings_explorer) (1.1.1)\n",
      "Requirement already satisfied: widgetsnbextension~=3.4.0 in ./ENV/lib/python3.7/site-packages (from ipywidgets->jupyter->biothings_explorer) (3.4.2)\n",
      "Requirement already satisfied: prompt-toolkit<2.1.0,>=2.0.0 in ./ENV/lib/python3.7/site-packages (from jupyter-console->jupyter->biothings_explorer) (2.0.9)\n",
      "Requirement already satisfied: webencodings in ./ENV/lib/python3.7/site-packages (from bleach->nbconvert->notebook==5.7.5->biothings_explorer) (0.5.1)\n",
      "Requirement already satisfied: pexpect; sys_platform != \"win32\" in ./ENV/lib/python3.7/site-packages (from ipython>=5.0.0->ipykernel->notebook==5.7.5->biothings_explorer) (4.7.0)\n",
      "Requirement already satisfied: appnope; sys_platform == \"darwin\" in ./ENV/lib/python3.7/site-packages (from ipython>=5.0.0->ipykernel->notebook==5.7.5->biothings_explorer) (0.1.0)\n",
      "Requirement already satisfied: pickleshare in ./ENV/lib/python3.7/site-packages (from ipython>=5.0.0->ipykernel->notebook==5.7.5->biothings_explorer) (0.7.5)\n",
      "Requirement already satisfied: backcall in ./ENV/lib/python3.7/site-packages (from ipython>=5.0.0->ipykernel->notebook==5.7.5->biothings_explorer) (0.1.0)\n",
      "Requirement already satisfied: jedi>=0.10 in ./ENV/lib/python3.7/site-packages (from ipython>=5.0.0->ipykernel->notebook==5.7.5->biothings_explorer) (0.13.3)\n",
      "Requirement already satisfied: wcwidth in ./ENV/lib/python3.7/site-packages (from prompt-toolkit<2.1.0,>=2.0.0->jupyter-console->jupyter->biothings_explorer) (0.1.7)\n",
      "Requirement already satisfied: parso>=0.3.0 in ./ENV/lib/python3.7/site-packages (from jedi>=0.10->ipython>=5.0.0->ipykernel->notebook==5.7.5->biothings_explorer) (0.4.0)\n",
      "\u001b[33mWARNING: You are using pip version 19.1.1, however version 20.0.2 is available.\n",
      "You should consider upgrading via the 'pip install --upgrade pip' command.\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "!pip install git+https://github.com/biothings/biothings_explorer#egg=biothings_explorer"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, import the relevant modules:\n",
    "\n",
    "* **Hint**: Find corresponding bio-entity representation used in BioThings Explorer based on user input (could be any database IDs, symbols, names)\n",
    "* **FindConnection**: Find intermediate bio-entities which connects user specified input and output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import modules from biothings_explorer\n",
    "from biothings_explorer.hint import Hint\n",
    "from biothings_explorer.user_query_dispatcher import FindConnection"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 1: Find representation of \"hyperphenylalaninemia\" in BTE\n",
    "\n",
    "In this step, BioThings Explorer translates our query string \"hyperphenylalaninemia\"  into BioThings objects, which contain mappings to many common identifiers.  Generally, the top result returned by the `Hint` module will be the correct item, but you should confirm that using the identifiers shown.\n",
    "\n",
    "Search terms can correspond to any child of [BiologicalEntity](https://biolink.github.io/biolink-model/docs/BiologicalEntity.html) from the [Biolink Model](https://biolink.github.io/biolink-model/docs/), including `DiseaseOrPhenotypicFeature` (e.g., \"lupus\"), `ChemicalSubstance` (e.g., \"acetaminophen\"), `Gene` (e.g., \"CDK2\"), `BiologicalProcess` (e.g., \"T cell differentiation\"), and `Pathway` (e.g., \"Citric acid cycle\")."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'MONDO': 'MONDO:0016543',\n",
       " 'UMLS': 'C0751435',\n",
       " 'name': 'hyperphenylalaninemia',\n",
       " 'ORPHANET': '238583',\n",
       " 'primary': {'identifier': 'MONDO',\n",
       "  'cls': 'Disease',\n",
       "  'value': 'MONDO:0016543'},\n",
       " 'display': 'MONDO(MONDO:0016543) ORPHANET(238583) UMLS(C0751435) name(hyperphenylalaninemia)',\n",
       " 'type': 'Disease'}"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ht = Hint()\n",
    "# find all potential representations of hyperphenylalaninemia\n",
    "hyperphenylalaninemia_hint = ht.query(\"hyperphenylalaninemia\")\n",
    "# select the correct representation of hyperphenylalaninemia\n",
    "hyperphenylalaninemia = hyperphenylalaninemia_hint['Disease'][0]\n",
    "hyperphenylalaninemia"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 2: Find drugs that are associated with genes which invovled in hyperphenylalaninemia\n",
    "\n",
    "In this section, we find all paths in the knowledge graph that connect hyperphenylalaninemia to any entity that is a chemical compound.  To do that, we will use `FindConnection`.  This class is a convenient wrapper around two advanced functions for **query path planning** and **query path execution**. More advanced features for both query path planning and query path execution are in development and will be documented in the coming months. \n",
    "\n",
    "The parameters for `FindConnection` are described below:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on function __init__ in module biothings_explorer.user_query_dispatcher:\n",
      "\n",
      "__init__(self, input_obj, output_obj, intermediate_nodes, registry=None)\n",
      "    Find relationships in the Knowledge Graph between an Input Object and an Output Object.\n",
      "    \n",
      "    Args:\n",
      "        input_obj (required): must be an object returned from Hint corresponding to a specific biomedical entity.\n",
      "                            Examples: \n",
      "                Hint().query(\"Fanconi anemia\")['DiseaseOrPhenotypicFeature'][0]\n",
      "                Hint().query(\"acetaminophen\")['ChemicalSubstance'][0]\n",
      "    \n",
      "        output_obj (required): must EITHER be an object returned from Hint corresponding to a specific biomedical\n",
      "                            entity, OR be a string or list of strings corresponding to Biolink Entity classes.\n",
      "                            Examples:\n",
      "                Hint().query(\"acetaminophen\")['ChemicalSubstance'][0]\n",
      "                'Gene'\n",
      "                ['Gene','ChemicalSubstance']\n",
      "    \n",
      "        intermediate_nodes (required): the semantic type(s) of the intermediate node(s).  Examples:\n",
      "                None                         : no intermediate node, find direct connections only\n",
      "                []                           : no intermediate node, find direct connections only\n",
      "                ['BiologicalEntity']         : one intermediate node of any semantic type\n",
      "                ['Gene']                     : one intermediate node that must be a Gene\n",
      "                [('Gene','Pathway')]         : one intermediate node that must be a Gene or a Pathway\n",
      "                ['Gene','Pathway']           : two intermediate nodes, first must be a Gene, second must be a Pathway.\n",
      "                ['Gene',('Pathway','Gene')]  : two intermediate nodes, first must be a Gene, second must be a Pathway or Gene.\n",
      "                                                **NOTE**: queries with more than one intermediate node are currently not supported\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(FindConnection.__init__)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here, we formulate a `FindConnection` query with \"hyperphenylalaninemia\" as the `input_ojb`, \"ChemicalSubstance\" as the `output_obj` (which corresponds to a Biolink Entity type).  We further specify with the `intermediate_nodes` parameter that we are looking for paths joining hyperphenylalaninemia and chemical compounds with *one* intermediate node that is a Gene.  (The ability to search for longer reasoning paths that include additional intermediate nodes will be added shortly.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "fc = FindConnection(input_obj=hyperphenylalaninemia, output_obj='ChemicalSubstance', intermediate_nodes=['Gene'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We next execute the `connect` method, which performs the **query path planning** and **query path execution** process.  In short, BioThings Explorer is deconstructing the query into individual API calls, executing those API calls, then assembling the results.\n",
    "\n",
    "A verbose log of this process is displayed below:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==========\n",
      "========== QUERY PARAMETER SUMMARY ==========\n",
      "==========\n",
      "\n",
      "BTE will find paths that join 'hyperphenylalaninemia' and 'ChemicalSubstance'.                   Paths will have 1 intermediate node.\n",
      "\n",
      "Intermediate node #1 will have these type constraints: Gene\n",
      "\n",
      "\n",
      "\n",
      "========== QUERY #1 -- fetch all Gene entities linked to hyperphenylalaninemia ==========\n",
      "==========\n",
      "\n",
      "==== Step #1: Query path planning ====\n",
      "\n",
      "Because hyperphenylalaninemia is of type 'Disease', BTE will query our meta-KG for APIs that can take 'Disease' as input and 'Gene' as output\n",
      "\n",
      "BTE found 7 apis:\n",
      "\n",
      "API 1. hetio(1 API call)\n",
      "API 2. scigraph(1 API call)\n",
      "API 3. semmed_disease(15 API calls)\n",
      "API 4. pharos(1 API call)\n",
      "API 5. scibite(1 API call)\n",
      "API 6. biolink(1 API call)\n",
      "API 7. mydisease(1 API call)\n",
      "\n",
      "\n",
      "==== Step #2: Query path execution ====\n",
      "NOTE: API requests are dispatched in parallel, so the list of APIs below is ordered by query time.\n",
      "\n",
      "API 7.1: http://mydisease.info/v1/query?fields=disgenet.genes_related_to_disease.gene_id (POST -d q=C0751435,C0751436&scopes=mondo.xrefs.umls, disgenet.xrefs.umls)\n",
      "API 3.1: https://biothings.ncats.io/semmed/query?fields=affects (POST -d q=C0751435,C0751436&scopes=umls)\n",
      "API 3.14: https://biothings.ncats.io/semmed/query?fields=caused_by (POST -d q=C0751435,C0751436&scopes=umls)\n",
      "API 3.13: https://biothings.ncats.io/semmed/query?fields=disrupts (POST -d q=C0751435,C0751436&scopes=umls)\n",
      "API 3.9: https://biothings.ncats.io/semmed/query?fields=disrupted_by (POST -d q=C0751435,C0751436&scopes=umls)\n",
      "API 3.15: https://biothings.ncats.io/semmed/query?fields=negatively_regulated_by (POST -d q=C0751435,C0751436&scopes=umls)\n",
      "API 3.2: https://biothings.ncats.io/semmed/query?fields=coexists_with (POST -d q=C0751435,C0751436&scopes=umls)\n",
      "API 3.8: https://biothings.ncats.io/semmed/query?fields=derives_from (POST -d q=C0751435,C0751436&scopes=umls)\n",
      "API 3.6: https://biothings.ncats.io/semmed/query?fields=treated_by (POST -d q=C0751435,C0751436&scopes=umls)\n",
      "API 3.10: https://biothings.ncats.io/semmed/query?fields=affected_by (POST -d q=C0751435,C0751436&scopes=umls)\n",
      "API 3.11: https://biothings.ncats.io/semmed/query?fields=negatively_regulates (POST -d q=C0751435,C0751436&scopes=umls)\n",
      "API 3.5: https://biothings.ncats.io/semmed/query?fields=physically_interacts_with (POST -d q=C0751435,C0751436&scopes=umls)\n",
      "API 3.7: https://biothings.ncats.io/semmed/query?fields=positively_regulated_by (POST -d q=C0751435,C0751436&scopes=umls)\n",
      "API 3.4: https://biothings.ncats.io/semmed/query?fields=positively_regulates (POST -d q=C0751435,C0751436&scopes=umls)\n",
      "API 3.3: https://biothings.ncats.io/semmed/query?fields=prevented_by (POST -d q=C0751435,C0751436&scopes=umls)\n",
      "API 3.12: https://biothings.ncats.io/semmed/query?fields=related_to (POST -d q=C0751435,C0751436&scopes=umls)\n",
      "API 6.1: https://api.monarchinitiative.org/api/bioentity/disease/MONDO:0016543/genes?rows=200\n",
      "API 5.1: https://automat.renci.org/cord19_scibite_v2/disease/gene/MONDO:0016543\n",
      "API 2.1: https://automat.renci.org/cord19_scigraph_v2/disease/gene/MONDO:0016543\n",
      "API 4.1: https://automat.renci.org/pharos/disease/gene/MONDO:0016543\n",
      "API 1.1: https://automat.renci.org/hetio/disease/gene/MONDO:0016543\n",
      "\n",
      "\n",
      "==== Step #3: Output normalization ====\n",
      "\n",
      "API 3.1 semmed_disease: No hits\n",
      "API 3.2 semmed_disease: No hits\n",
      "API 3.3 semmed_disease: No hits\n",
      "API 3.4 semmed_disease: No hits\n",
      "API 3.5 semmed_disease: No hits\n",
      "API 3.6 semmed_disease: 6 hits\n",
      "API 3.7 semmed_disease: No hits\n",
      "API 4.1 pharos: 1 hits\n",
      "API 3.8 semmed_disease: No hits\n",
      "API 2.1 scigraph: No hits\n",
      "API 3.9 semmed_disease: 3 hits\n",
      "API 7.1 mydisease: 10 hits\n",
      "API 3.10 semmed_disease: No hits\n",
      "API 5.1 scibite: 1 hits\n",
      "API 1.1 hetio: No hits\n",
      "API 3.11 semmed_disease: No hits\n",
      "API 3.12 semmed_disease: 14 hits\n",
      "API 3.13 semmed_disease: No hits\n",
      "API 3.14 semmed_disease: 17 hits\n",
      "API 6.1 biolink: 4 hits\n",
      "API 3.15 semmed_disease: No hits\n",
      "\n",
      "After id-to-object translation, BTE retrieved 33 unique objects.\n",
      "\n",
      "\n",
      "========== QUERY #2.1 -- fetch all ChemicalSubstance entities linked to Gene entites ==========\n",
      "==========\n",
      "\n",
      "==== Step #1: Query path planning ====\n",
      "\n",
      "Because None is of type 'Gene', BTE will query our meta-KG for APIs that can take 'Gene' as input and 'ChemicalSubstance' as output\n",
      "\n",
      "BTE found 10 apis:\n",
      "\n",
      "API 1. semmed_gene(10 API calls)\n",
      "API 2. scigraph(16 API calls)\n",
      "API 3. hmdb(16 API calls)\n",
      "API 4. mychem(3 API calls)\n",
      "API 5. dgidb(16 API calls)\n",
      "API 6. pharos(16 API calls)\n",
      "API 7. scibite(16 API calls)\n",
      "API 8. opentarget(16 API calls)\n",
      "API 9. chembio(16 API calls)\n",
      "API 10. cord_gene(1 API call)\n",
      "\n",
      "\n",
      "==== Step #2: Query path execution ====\n",
      "NOTE: API requests are dispatched in parallel, so the list of APIs below is ordered by query time.\n",
      "\n",
      "API 7.15: https://platform-api.opentargets.io/v3/platform/public/evidence/filter?target=ENSG00000145730&datasource=chembl&size=100&fields=drug\n",
      "API 7.8: https://platform-api.opentargets.io/v3/platform/public/evidence/filter?target=ENSG00000171759&datasource=chembl&size=100&fields=drug\n",
      "API 7.7: https://platform-api.opentargets.io/v3/platform/public/evidence/filter?target=ENSG00000166228&datasource=chembl&size=100&fields=drug\n",
      "API 7.5: https://platform-api.opentargets.io/v3/platform/public/evidence/filter?target=ENSG00000137880&datasource=chembl&size=100&fields=drug\n",
      "API 7.1: https://platform-api.opentargets.io/v3/platform/public/evidence/filter?target=ENSG00000151552&datasource=chembl&size=100&fields=drug\n",
      "API 7.13: https://platform-api.opentargets.io/v3/platform/public/evidence/filter?target=ENSG00000148602&datasource=chembl&size=100&fields=drug\n",
      "API 7.6: https://platform-api.opentargets.io/v3/platform/public/evidence/filter?target=ENSG00000198074&datasource=chembl&size=100&fields=drug\n",
      "API 7.11: https://platform-api.opentargets.io/v3/platform/public/evidence/filter?target=ENSG00000141837&datasource=chembl&size=100&fields=drug\n",
      "API 7.3: https://platform-api.opentargets.io/v3/platform/public/evidence/filter?target=ENSG00000131979&datasource=chembl&size=100&fields=drug\n",
      "API 7.12: https://platform-api.opentargets.io/v3/platform/public/evidence/filter?target=ENSG00000150787&datasource=chembl&size=100&fields=drug\n",
      "API 7.10: https://platform-api.opentargets.io/v3/platform/public/evidence/filter?target=ENSG00000170606&datasource=chembl&size=100&fields=drug\n",
      "API 7.16: https://platform-api.opentargets.io/v3/platform/public/evidence/filter?target=ENSG00000135775&datasource=chembl&size=100&fields=drug\n",
      "API 7.4: https://platform-api.opentargets.io/v3/platform/public/evidence/filter?target=ENSG00000142192&datasource=chembl&size=100&fields=drug\n",
      "API 5.13: http://dgidb.genome.wustl.edu/api/v2/interactions.json?genes=AKR1B10\n",
      "API 5.7: http://dgidb.genome.wustl.edu/api/v2/interactions.json?genes=CACNA1A\n",
      "API 5.14: http://dgidb.genome.wustl.edu/api/v2/interactions.json?genes=PAH\n",
      "API 5.8: http://dgidb.genome.wustl.edu/api/v2/interactions.json?genes=PTS\n",
      "API 5.9: http://dgidb.genome.wustl.edu/api/v2/interactions.json?genes=PAM\n",
      "API 5.11: http://dgidb.genome.wustl.edu/api/v2/interactions.json?genes=HSPA4\n",
      "API 5.2: http://dgidb.genome.wustl.edu/api/v2/interactions.json?genes=GCHFR\n",
      "API 5.3: http://dgidb.genome.wustl.edu/api/v2/interactions.json?genes=GCH1\n",
      "API 5.5: http://dgidb.genome.wustl.edu/api/v2/interactions.json?genes=QDPR\n",
      "API 5.6: http://dgidb.genome.wustl.edu/api/v2/interactions.json?genes=LRIT1\n",
      "API 5.16: http://dgidb.genome.wustl.edu/api/v2/interactions.json?genes=COG2\n",
      "API 5.1: http://dgidb.genome.wustl.edu/api/v2/interactions.json?genes=PCBD1\n",
      "API 5.12: http://dgidb.genome.wustl.edu/api/v2/interactions.json?genes=APP\n",
      "API 6.16: https://automat.renci.org/cord19_scibite_v2/gene/chemical_substance/NCBIGene:338\n",
      "API 8.1: https://automat.renci.org/chembio/gene/chemical_substance/NCBIGene:57016\n",
      "API 2.7: https://automat.renci.org/cord19_scigraph_v2/gene/chemical_substance/NCBIGene:3308\n",
      "API 4.8: https://automat.renci.org/pharos/gene/chemical_substance/NCBIGene:2644\n",
      "API 4.7: https://automat.renci.org/pharos/gene/chemical_substance/NCBIGene:3308\n",
      "API 6.1: https://automat.renci.org/cord19_scibite_v2/gene/chemical_substance/NCBIGene:57016\n",
      "API 6.13: https://automat.renci.org/cord19_scibite_v2/gene/chemical_substance/NCBIGene:2643\n",
      "API 6.10: https://automat.renci.org/cord19_scibite_v2/gene/chemical_substance/NCBIGene:351\n",
      "API 4.2: https://automat.renci.org/pharos/gene/chemical_substance/NCBIGene:5860\n",
      "API 6.6: https://automat.renci.org/cord19_scibite_v2/gene/chemical_substance/NCBIGene:5805\n",
      "API 6.3: https://automat.renci.org/cord19_scibite_v2/gene/chemical_substance/NCBIGene:773\n",
      "API 6.15: https://automat.renci.org/cord19_scibite_v2/gene/chemical_substance/NCBIGene:5066\n",
      "API 2.5: https://automat.renci.org/cord19_scigraph_v2/gene/chemical_substance/NCBIGene:5092\n",
      "API 2.14: https://automat.renci.org/cord19_scigraph_v2/gene/chemical_substance/NCBIGene:26103\n",
      "API 2.2: https://automat.renci.org/cord19_scigraph_v2/gene/chemical_substance/NCBIGene:5860\n",
      "API 4.1: https://automat.renci.org/pharos/gene/chemical_substance/NCBIGene:57016\n",
      "API 2.11: https://automat.renci.org/cord19_scigraph_v2/gene/chemical_substance/NCBIGene:3034\n",
      "API 4.5: https://automat.renci.org/pharos/gene/chemical_substance/NCBIGene:5092\n",
      "API 5.15: http://dgidb.genome.wustl.edu/api/v2/interactions.json?genes=SHCBP1\n",
      "API 6.4: https://automat.renci.org/cord19_scibite_v2/gene/chemical_substance/NCBIGene:22796\n",
      "API 2.8: https://automat.renci.org/cord19_scigraph_v2/gene/chemical_substance/NCBIGene:2644\n",
      "API 2.12: https://automat.renci.org/cord19_scigraph_v2/gene/chemical_substance/NCBIGene:5053\n",
      "API 5.4: http://dgidb.genome.wustl.edu/api/v2/interactions.json?genes=HAL\n",
      "API 3.2: https://mychem.info/v1/query?fields=drugbank.id&size=250 (POST -d q=PCBD1,GCHFR,GCH1,HAL,QDPR,LRIT1,CACNA1A,PTS,PAM,APOB,HSPA4,APP,AKR1B10,PAH,SHCBP1,COG2&scopes=drugbank.enzymes.gene_name)\n",
      "API 3.1: https://mychem.info/v1/query?fields=drugbank.id&size=250 (POST -d q=PCBD1,GCHFR,GCH1,HAL,QDPR,LRIT1,CACNA1A,PTS,PAM,APOB,HSPA4,APP,AKR1B10,PAH,SHCBP1,COG2&scopes=drugbank.targets.gene_name)\n",
      "API 6.9: https://automat.renci.org/cord19_scibite_v2/gene/chemical_substance/NCBIGene:79801\n",
      "API 6.7: https://automat.renci.org/cord19_scibite_v2/gene/chemical_substance/NCBIGene:3308\n",
      "API 5.10: http://dgidb.genome.wustl.edu/api/v2/interactions.json?genes=APOB\n",
      "API 7.2: https://platform-api.opentargets.io/v3/platform/public/evidence/filter?target=ENSG00000084674&datasource=chembl&size=100&fields=drug\n",
      "API 2.9: https://automat.renci.org/cord19_scigraph_v2/gene/chemical_substance/NCBIGene:79801\n",
      "API 6.2: https://automat.renci.org/cord19_scibite_v2/gene/chemical_substance/NCBIGene:5860\n",
      "API 6.5: https://automat.renci.org/cord19_scibite_v2/gene/chemical_substance/NCBIGene:5092\n",
      "API 6.14: https://automat.renci.org/cord19_scibite_v2/gene/chemical_substance/NCBIGene:26103\n",
      "API 6.8: https://automat.renci.org/cord19_scibite_v2/gene/chemical_substance/NCBIGene:2644\n",
      "API 2.16: https://automat.renci.org/cord19_scigraph_v2/gene/chemical_substance/NCBIGene:338\n",
      "API 6.11: https://automat.renci.org/cord19_scibite_v2/gene/chemical_substance/NCBIGene:3034\n",
      "API 7.9: https://platform-api.opentargets.io/v3/platform/public/evidence/filter?target=ENSG00000171241&datasource=chembl&size=100&fields=drug\n",
      "API 4.6: https://automat.renci.org/pharos/gene/chemical_substance/NCBIGene:5805\n",
      "API 6.12: https://automat.renci.org/cord19_scibite_v2/gene/chemical_substance/NCBIGene:5053\n",
      "API 4.12: https://automat.renci.org/pharos/gene/chemical_substance/NCBIGene:5053\n",
      "API 2.4: https://automat.renci.org/cord19_scigraph_v2/gene/chemical_substance/NCBIGene:22796\n",
      "API 7.14: https://platform-api.opentargets.io/v3/platform/public/evidence/filter?target=ENSG00000084110&datasource=chembl&size=100&fields=drug\n",
      "API 3.3: https://mychem.info/v1/query?fields=chembl.molecule_chembl_id&size=250 (POST -d q=PCBD1,GCHFR,GCH1,HAL,QDPR,LRIT1,CACNA1A,PTS,PAM,APOB,HSPA4,APP,AKR1B10,PAH,SHCBP1,COG2&scopes=drugcentral.bioactivity.uniprot.gene_symbol)\n",
      "API 2.1: https://automat.renci.org/cord19_scigraph_v2/gene/chemical_substance/NCBIGene:57016\n",
      "API 2.13: https://automat.renci.org/cord19_scigraph_v2/gene/chemical_substance/NCBIGene:2643\n",
      "API 4.9: https://automat.renci.org/pharos/gene/chemical_substance/NCBIGene:79801\n",
      "API 8.2: https://automat.renci.org/chembio/gene/chemical_substance/NCBIGene:5860\n",
      "API 4.3: https://automat.renci.org/pharos/gene/chemical_substance/NCBIGene:773\n",
      "API 9.6: https://automat.renci.org/hmdb/gene/chemical_substance/NCBIGene:5805\n",
      "API 8.7: https://automat.renci.org/chembio/gene/chemical_substance/NCBIGene:3308\n",
      "API 4.13: https://automat.renci.org/pharos/gene/chemical_substance/NCBIGene:2643\n",
      "API 4.4: https://automat.renci.org/pharos/gene/chemical_substance/NCBIGene:22796\n",
      "API 2.15: https://automat.renci.org/cord19_scigraph_v2/gene/chemical_substance/NCBIGene:5066\n",
      "API 2.3: https://automat.renci.org/cord19_scigraph_v2/gene/chemical_substance/NCBIGene:773\n",
      "API 4.14: https://automat.renci.org/pharos/gene/chemical_substance/NCBIGene:26103\n",
      "API 4.11: https://automat.renci.org/pharos/gene/chemical_substance/NCBIGene:3034\n",
      "API 2.10: https://automat.renci.org/cord19_scigraph_v2/gene/chemical_substance/NCBIGene:351\n",
      "API 4.15: https://automat.renci.org/pharos/gene/chemical_substance/NCBIGene:5066\n",
      "API 8.3: https://automat.renci.org/chembio/gene/chemical_substance/NCBIGene:773\n",
      "API 8.9: https://automat.renci.org/chembio/gene/chemical_substance/NCBIGene:79801\n",
      "API 4.16: https://automat.renci.org/pharos/gene/chemical_substance/NCBIGene:338\n",
      "API 8.8: https://automat.renci.org/chembio/gene/chemical_substance/NCBIGene:2644\n",
      "API 8.4: https://automat.renci.org/chembio/gene/chemical_substance/NCBIGene:22796\n",
      "API 8.15: https://automat.renci.org/chembio/gene/chemical_substance/NCBIGene:5066\n",
      "API 8.5: https://automat.renci.org/chembio/gene/chemical_substance/NCBIGene:5092\n",
      "API 8.6: https://automat.renci.org/chembio/gene/chemical_substance/NCBIGene:5805\n",
      "API 8.10: https://automat.renci.org/chembio/gene/chemical_substance/NCBIGene:351\n",
      "API 2.6: https://automat.renci.org/cord19_scigraph_v2/gene/chemical_substance/NCBIGene:5805\n",
      "API 8.11: https://automat.renci.org/chembio/gene/chemical_substance/NCBIGene:3034\n",
      "API 8.14: https://automat.renci.org/chembio/gene/chemical_substance/NCBIGene:26103\n",
      "API 8.13: https://automat.renci.org/chembio/gene/chemical_substance/NCBIGene:2643\n",
      "API 4.10: https://automat.renci.org/pharos/gene/chemical_substance/NCBIGene:351\n",
      "API 9.4: https://automat.renci.org/hmdb/gene/chemical_substance/NCBIGene:22796\n",
      "API 8.12: https://automat.renci.org/chembio/gene/chemical_substance/NCBIGene:5053\n",
      "API 8.16: https://automat.renci.org/chembio/gene/chemical_substance/NCBIGene:338\n",
      "API 9.1: https://automat.renci.org/hmdb/gene/chemical_substance/NCBIGene:57016\n",
      "API 10.1: https://biothings.ncats.io/cord_gene/query?fields=associated_with (POST -d q=4193,6546,5237,9689,1388,603,4194,382,4806,29547,8646,8596,9752,23404,620,8582&scopes=hgnc)\n",
      "API 9.7: https://automat.renci.org/hmdb/gene/chemical_substance/NCBIGene:3308\n",
      "API 9.8: https://automat.renci.org/hmdb/gene/chemical_substance/NCBIGene:2644\n",
      "API 9.9: https://automat.renci.org/hmdb/gene/chemical_substance/NCBIGene:79801\n",
      "API 9.3: https://automat.renci.org/hmdb/gene/chemical_substance/NCBIGene:773\n",
      "API 9.5: https://automat.renci.org/hmdb/gene/chemical_substance/NCBIGene:5092\n",
      "API 9.2: https://automat.renci.org/hmdb/gene/chemical_substance/NCBIGene:5860\n",
      "API 9.12: https://automat.renci.org/hmdb/gene/chemical_substance/NCBIGene:5053\n",
      "API 9.10: https://automat.renci.org/hmdb/gene/chemical_substance/NCBIGene:351\n",
      "API 9.11: https://automat.renci.org/hmdb/gene/chemical_substance/NCBIGene:3034\n",
      "API 9.14: https://automat.renci.org/hmdb/gene/chemical_substance/NCBIGene:26103\n",
      "API 9.15: https://automat.renci.org/hmdb/gene/chemical_substance/NCBIGene:5066\n",
      "API 9.13: https://automat.renci.org/hmdb/gene/chemical_substance/NCBIGene:2643\n",
      "API 9.16: https://automat.renci.org/hmdb/gene/chemical_substance/NCBIGene:338\n",
      "API 1.1: https://biothings.ncats.io/semmedgene/query?fields=negatively_regulates (POST -d q=C0017361,C0018293,C1412323,C0072774,C1364818,C1415465,C1335144,C1539766,C0314604,C1416822,C0031456,C0012312,C1710338,C1428072,C0074350,C0002085,C0014442,C1418251,C1418296,C1456413,C0041491,C0284927,C1419129,C1705543,C1415012,C1419182,C1412471,C1415760,C1415013,C0020239,C0030016,C1456357,C0017337&scopes=umls)\n",
      "API 1.6: https://biothings.ncats.io/semmedgene/query?fields=positively_regulated_by (POST -d q=C0017361,C0018293,C1412323,C0072774,C1364818,C1415465,C1335144,C1539766,C0314604,C1416822,C0031456,C0012312,C1710338,C1428072,C0074350,C0002085,C0014442,C1418251,C1418296,C1456413,C0041491,C0284927,C1419129,C1705543,C1415012,C1419182,C1412471,C1415760,C1415013,C0020239,C0030016,C1456357,C0017337&scopes=umls)\n",
      "API 1.4: https://biothings.ncats.io/semmedgene/query?fields=affected_by (POST -d q=C0017361,C0018293,C1412323,C0072774,C1364818,C1415465,C1335144,C1539766,C0314604,C1416822,C0031456,C0012312,C1710338,C1428072,C0074350,C0002085,C0014442,C1418251,C1418296,C1456413,C0041491,C0284927,C1419129,C1705543,C1415012,C1419182,C1412471,C1415760,C1415013,C0020239,C0030016,C1456357,C0017337&scopes=umls)\n",
      "API 1.9: https://biothings.ncats.io/semmedgene/query?fields=disrupted_by (POST -d q=C0017361,C0018293,C1412323,C0072774,C1364818,C1415465,C1335144,C1539766,C0314604,C1416822,C0031456,C0012312,C1710338,C1428072,C0074350,C0002085,C0014442,C1418251,C1418296,C1456413,C0041491,C0284927,C1419129,C1705543,C1415012,C1419182,C1412471,C1415760,C1415013,C0020239,C0030016,C1456357,C0017337&scopes=umls)\n",
      "API 1.8: https://biothings.ncats.io/semmedgene/query?fields=disrupts (POST -d q=C0017361,C0018293,C1412323,C0072774,C1364818,C1415465,C1335144,C1539766,C0314604,C1416822,C0031456,C0012312,C1710338,C1428072,C0074350,C0002085,C0014442,C1418251,C1418296,C1456413,C0041491,C0284927,C1419129,C1705543,C1415012,C1419182,C1412471,C1415760,C1415013,C0020239,C0030016,C1456357,C0017337&scopes=umls)\n",
      "API 1.7: https://biothings.ncats.io/semmedgene/query?fields=negatively_regulated_by (POST -d q=C0017361,C0018293,C1412323,C0072774,C1364818,C1415465,C1335144,C1539766,C0314604,C1416822,C0031456,C0012312,C1710338,C1428072,C0074350,C0002085,C0014442,C1418251,C1418296,C1456413,C0041491,C0284927,C1419129,C1705543,C1415012,C1419182,C1412471,C1415760,C1415013,C0020239,C0030016,C1456357,C0017337&scopes=umls)\n",
      "API 1.10: https://biothings.ncats.io/semmedgene/query?fields=positively_regulates (POST -d q=C0017361,C0018293,C1412323,C0072774,C1364818,C1415465,C1335144,C1539766,C0314604,C1416822,C0031456,C0012312,C1710338,C1428072,C0074350,C0002085,C0014442,C1418251,C1418296,C1456413,C0041491,C0284927,C1419129,C1705543,C1415012,C1419182,C1412471,C1415760,C1415013,C0020239,C0030016,C1456357,C0017337&scopes=umls)\n",
      "API 1.5: https://biothings.ncats.io/semmedgene/query?fields=related_to (POST -d q=C0017361,C0018293,C1412323,C0072774,C1364818,C1415465,C1335144,C1539766,C0314604,C1416822,C0031456,C0012312,C1710338,C1428072,C0074350,C0002085,C0014442,C1418251,C1418296,C1456413,C0041491,C0284927,C1419129,C1705543,C1415012,C1419182,C1412471,C1415760,C1415013,C0020239,C0030016,C1456357,C0017337&scopes=umls)\n",
      "API 1.3: https://biothings.ncats.io/semmedgene/query?fields=affects (POST -d q=C0017361,C0018293,C1412323,C0072774,C1364818,C1415465,C1335144,C1539766,C0314604,C1416822,C0031456,C0012312,C1710338,C1428072,C0074350,C0002085,C0014442,C1418251,C1418296,C1456413,C0041491,C0284927,C1419129,C1705543,C1415012,C1419182,C1412471,C1415760,C1415013,C0020239,C0030016,C1456357,C0017337&scopes=umls)\n",
      "API 1.2: https://biothings.ncats.io/semmedgene/query?fields=physically_interacts_with (POST -d q=C0017361,C0018293,C1412323,C0072774,C1364818,C1415465,C1335144,C1539766,C0314604,C1416822,C0031456,C0012312,C1710338,C1428072,C0074350,C0002085,C0014442,C1418251,C1418296,C1456413,C0041491,C0284927,C1419129,C1705543,C1415012,C1419182,C1412471,C1415760,C1415013,C0020239,C0030016,C1456357,C0017337&scopes=umls)\n",
      "\n",
      "\n",
      "==== Step #3: Output normalization ====\n",
      "\n",
      "API 7.1 opentarget: No hits\n",
      "API 7.2 opentarget: 7 hits\n",
      "API 7.3 opentarget: No hits\n",
      "API 7.4 opentarget: 59 hits\n",
      "API 7.5 opentarget: No hits\n",
      "API 7.6 opentarget: No hits\n",
      "API 7.7 opentarget: No hits\n",
      "API 7.8 opentarget: 42 hits\n",
      "API 7.9 opentarget: No hits\n",
      "API 7.10 opentarget: No hits\n",
      "API 7.11 opentarget: 19 hits\n",
      "API 7.12 opentarget: No hits\n",
      "API 7.13 opentarget: No hits\n",
      "API 7.14 opentarget: No hits\n",
      "API 7.15 opentarget: No hits\n",
      "API 7.16 opentarget: No hits\n",
      "API 6.1 scibite: 1 hits\n",
      "API 6.2 scibite: No hits\n",
      "API 6.3 scibite: No hits\n",
      "API 6.4 scibite: No hits\n",
      "API 6.5 scibite: No hits\n",
      "API 6.6 scibite: No hits\n",
      "API 6.7 scibite: 1 hits\n",
      "API 6.8 scibite: 1 hits\n",
      "API 6.9 scibite: No hits\n",
      "API 6.10 scibite: 9 hits\n",
      "API 6.11 scibite: No hits\n",
      "API 6.12 scibite: 1 hits\n",
      "API 6.13 scibite: 1 hits\n",
      "API 6.14 scibite: No hits\n",
      "API 6.15 scibite: No hits\n",
      "API 6.16 scibite: 3 hits\n",
      "API 1.1 semmed_gene: 2007 hits\n",
      "API 1.2 semmed_gene: 15441 hits\n",
      "API 1.3 semmed_gene: 1 hits\n",
      "API 1.4 semmed_gene: 9 hits\n",
      "API 2.1 scigraph: No hits\n",
      "API 2.2 scigraph: No hits\n",
      "API 2.3 scigraph: No hits\n",
      "API 2.4 scigraph: No hits\n",
      "API 2.5 scigraph: No hits\n",
      "API 2.6 scigraph: 442 hits\n",
      "API 2.7 scigraph: No hits\n",
      "API 2.8 scigraph: No hits\n",
      "API 2.9 scigraph: No hits\n",
      "API 2.10 scigraph: 56 hits\n",
      "API 2.11 scigraph: 1 hits\n",
      "API 2.12 scigraph: 52 hits\n",
      "API 2.13 scigraph: No hits\n",
      "API 2.14 scigraph: No hits\n",
      "API 2.15 scigraph: 19 hits\n",
      "API 2.16 scigraph: 16 hits\n",
      "API 1.5 semmed_gene: No hits\n",
      "API 5.1 dgidb: 1 hits\n",
      "API 5.2 dgidb: No hits\n",
      "API 5.3 dgidb: 1 hits\n",
      "API 5.4 dgidb: 1 hits\n",
      "API 5.5 dgidb: 1 hits\n",
      "API 5.6 dgidb: No hits\n",
      "API 5.7 dgidb: 11 hits\n",
      "API 5.8 dgidb: No hits\n",
      "API 5.9 dgidb: 1 hits\n",
      "API 5.10 dgidb: 17 hits\n",
      "API 5.11 dgidb: 35 hits\n",
      "API 5.12 dgidb: 22 hits\n",
      "API 5.13 dgidb: 3 hits\n",
      "API 5.14 dgidb: 8 hits\n",
      "API 5.15 dgidb: No hits\n",
      "API 5.16 dgidb: No hits\n",
      "API 1.6 semmed_gene: 4883 hits\n",
      "API 3.1 mychem: 51 hits\n",
      "API 10.1 cord_gene: 723 hits\n",
      "API 1.7 semmed_gene: 6577 hits\n",
      "API 4.1 pharos: 44 hits\n",
      "API 4.2 pharos: 3 hits\n",
      "API 4.3 pharos: 1 hits\n",
      "API 4.4 pharos: No hits\n",
      "API 4.5 pharos: No hits\n",
      "API 4.6 pharos: No hits\n",
      "API 4.7 pharos: No hits\n",
      "API 4.8 pharos: No hits\n",
      "API 4.9 pharos: No hits\n",
      "API 4.10 pharos: 358 hits\n",
      "API 4.11 pharos: No hits\n",
      "API 4.12 pharos: 2 hits\n",
      "API 4.13 pharos: No hits\n",
      "API 4.14 pharos: No hits\n",
      "API 4.15 pharos: 5 hits\n",
      "API 4.16 pharos: 49 hits\n",
      "API 1.8 semmed_gene: 2 hits\n",
      "API 8.1 chembio: 5 hits\n",
      "API 8.2 chembio: 1 hits\n",
      "API 8.3 chembio: 1 hits\n",
      "API 8.4 chembio: No hits\n",
      "API 8.5 chembio: 2 hits\n",
      "API 8.6 chembio: 2 hits\n",
      "API 8.7 chembio: 6 hits\n",
      "API 8.8 chembio: 4 hits\n",
      "API 8.9 chembio: 1 hits\n",
      "API 8.10 chembio: 29 hits\n",
      "API 8.11 chembio: 3 hits\n",
      "API 8.12 chembio: 1 hits\n",
      "API 8.13 chembio: 4 hits\n",
      "API 8.14 chembio: No hits\n",
      "API 8.15 chembio: 7 hits\n",
      "API 8.16 chembio: 6 hits\n",
      "API 3.2 mychem: 4 hits\n",
      "API 1.9 semmed_gene: No hits\n",
      "API 9.1 hmdb: 18 hits\n",
      "API 9.2 hmdb: 11 hits\n",
      "API 9.3 hmdb: 4 hits\n",
      "API 9.4 hmdb: No hits\n",
      "API 9.5 hmdb: 5 hits\n",
      "API 9.6 hmdb: 6 hits\n",
      "API 9.7 hmdb: No hits\n",
      "API 9.8 hmdb: 2 hits\n",
      "API 9.9 hmdb: No hits\n",
      "API 9.10 hmdb: 5 hits\n",
      "API 9.11 hmdb: 3 hits\n",
      "API 9.12 hmdb: 11 hits\n",
      "API 9.13 hmdb: 8 hits\n",
      "API 9.14 hmdb: No hits\n",
      "API 9.15 hmdb: 5 hits\n",
      "API 9.16 hmdb: 4 hits\n",
      "API 3.3 mychem: 15 hits\n",
      "API 1.10 semmed_gene: 2314 hits\n",
      "\n",
      "After id-to-object translation, BTE retrieved 9449 unique objects.\n",
      "\n",
      "==========\n",
      "========== Final assembly of results ==========\n",
      "==========\n",
      "\n",
      "\n",
      "In the #1 query, BTE found 33 unique Gene nodes\n",
      "In the #2 query, BTE found 9449 unique ChemicalSubstance nodes\n"
     ]
    }
   ],
   "source": [
    "# set verbose=True will display all steps which BTE takes to find the connection\n",
    "fc.connect(verbose=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 3: Display and Filter results\n",
    "This section demonstrates post-query filtering done in Python. Later, more advanced filtering functions will be added to the **query path execution** module for interleaved filtering, thereby enabling longer query paths. More details to come...\n",
    "\n",
    "First, all matching paths can be exported to a data frame. Let's examine a sample of those results."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>input</th>\n",
       "      <th>input_type</th>\n",
       "      <th>pred1</th>\n",
       "      <th>pred1_source</th>\n",
       "      <th>pred1_api</th>\n",
       "      <th>pred1_pubmed</th>\n",
       "      <th>node1_type</th>\n",
       "      <th>node1_name</th>\n",
       "      <th>node1_id</th>\n",
       "      <th>pred2</th>\n",
       "      <th>pred2_source</th>\n",
       "      <th>pred2_api</th>\n",
       "      <th>pred2_pubmed</th>\n",
       "      <th>output_type</th>\n",
       "      <th>output_name</th>\n",
       "      <th>output_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>16504</td>\n",
       "      <td>HYPERPHENYLALANINAEMIA</td>\n",
       "      <td>Disease</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Disease API</td>\n",
       "      <td>10727395</td>\n",
       "      <td>Gene</td>\n",
       "      <td>C0014442</td>\n",
       "      <td>UMLS:C0014442</td>\n",
       "      <td>negatively_regulated_by</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Gene API</td>\n",
       "      <td>11839529,19443747</td>\n",
       "      <td>ChemicalSubstance</td>\n",
       "      <td>U 0126</td>\n",
       "      <td>name:U 0126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8209</td>\n",
       "      <td>HYPERPHENYLALANINAEMIA</td>\n",
       "      <td>Disease</td>\n",
       "      <td>related_to</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Disease API</td>\n",
       "      <td>9675155</td>\n",
       "      <td>Gene</td>\n",
       "      <td>C0017337</td>\n",
       "      <td>UMLS:C0017337</td>\n",
       "      <td>positively_regulated_by</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Gene API</td>\n",
       "      <td>21984199,26086415</td>\n",
       "      <td>ChemicalSubstance</td>\n",
       "      <td>ISOFLURANE</td>\n",
       "      <td>name:ISOFLURANE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>31798</td>\n",
       "      <td>HYPERPHENYLALANINAEMIA</td>\n",
       "      <td>Disease</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Disease API</td>\n",
       "      <td>9222755,9667588</td>\n",
       "      <td>Gene</td>\n",
       "      <td>C0017337</td>\n",
       "      <td>UMLS:C0017337</td>\n",
       "      <td>positively_regulated_by</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Gene API</td>\n",
       "      <td>17684401,18412988,17693008</td>\n",
       "      <td>ChemicalSubstance</td>\n",
       "      <td>MICRONUTRIENT</td>\n",
       "      <td>name:MICRONUTRIENT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1929</td>\n",
       "      <td>HYPERPHENYLALANINAEMIA</td>\n",
       "      <td>Disease</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Disease API</td>\n",
       "      <td>10874306</td>\n",
       "      <td>Gene</td>\n",
       "      <td>PTS</td>\n",
       "      <td>NCBIGene:5805</td>\n",
       "      <td>related_to</td>\n",
       "      <td>Translator Text Mining Provider</td>\n",
       "      <td>CORD Gene API</td>\n",
       "      <td>None</td>\n",
       "      <td>ChemicalSubstance</td>\n",
       "      <td>.BETA.-D-GLUCOPYRANOSE</td>\n",
       "      <td>name:.BETA.-D-GLUCOPYRANOSE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>10532</td>\n",
       "      <td>HYPERPHENYLALANINAEMIA</td>\n",
       "      <td>Disease</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Disease API</td>\n",
       "      <td>699146</td>\n",
       "      <td>Gene</td>\n",
       "      <td>C0030016</td>\n",
       "      <td>UMLS:C0030016</td>\n",
       "      <td>positively_regulated_by</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Gene API</td>\n",
       "      <td>11731036</td>\n",
       "      <td>ChemicalSubstance</td>\n",
       "      <td>GENISTEIN</td>\n",
       "      <td>name:GENISTEIN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>35512</td>\n",
       "      <td>HYPERPHENYLALANINAEMIA</td>\n",
       "      <td>Disease</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Disease API</td>\n",
       "      <td>10727395</td>\n",
       "      <td>Gene</td>\n",
       "      <td>C0014442</td>\n",
       "      <td>UMLS:C0014442</td>\n",
       "      <td>positively_regulates</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Gene API</td>\n",
       "      <td>4291334,7358701,11980911</td>\n",
       "      <td>ChemicalSubstance</td>\n",
       "      <td>DEOXYCHOLATE</td>\n",
       "      <td>name:DEOXYCHOLATE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>40287</td>\n",
       "      <td>HYPERPHENYLALANINAEMIA</td>\n",
       "      <td>Disease</td>\n",
       "      <td>treated_by</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Disease API</td>\n",
       "      <td>8841415</td>\n",
       "      <td>Gene</td>\n",
       "      <td>PTS</td>\n",
       "      <td>NCBIGene:5805</td>\n",
       "      <td>related_to</td>\n",
       "      <td>scigraph</td>\n",
       "      <td>Automat CORD19 Scigraph API</td>\n",
       "      <td>None</td>\n",
       "      <td>ChemicalSubstance</td>\n",
       "      <td>CHLORAMBUCIL</td>\n",
       "      <td>name:CHLORAMBUCIL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>40027</td>\n",
       "      <td>HYPERPHENYLALANINAEMIA</td>\n",
       "      <td>Disease</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Disease API</td>\n",
       "      <td>15133828</td>\n",
       "      <td>Gene</td>\n",
       "      <td>AKR1B10</td>\n",
       "      <td>NCBIGene:57016</td>\n",
       "      <td>negatively_regulates</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Gene API</td>\n",
       "      <td>28322780</td>\n",
       "      <td>ChemicalSubstance</td>\n",
       "      <td>TIAPROFENIC ACID</td>\n",
       "      <td>name:TIAPROFENIC ACID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>22054</td>\n",
       "      <td>HYPERPHENYLALANINAEMIA</td>\n",
       "      <td>Disease</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Disease API</td>\n",
       "      <td>28794131</td>\n",
       "      <td>Gene</td>\n",
       "      <td>HSPA4</td>\n",
       "      <td>NCBIGene:3308</td>\n",
       "      <td>positively_regulated_by</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Gene API</td>\n",
       "      <td>26307981</td>\n",
       "      <td>ChemicalSubstance</td>\n",
       "      <td>MALONALDEHYDE</td>\n",
       "      <td>name:MALONALDEHYDE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>45089</td>\n",
       "      <td>HYPERPHENYLALANINAEMIA</td>\n",
       "      <td>Disease</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Disease API</td>\n",
       "      <td>1301201,8892014,10495930,10495930</td>\n",
       "      <td>Gene</td>\n",
       "      <td>PAH</td>\n",
       "      <td>NCBIGene:5053</td>\n",
       "      <td>related_to</td>\n",
       "      <td>Translator Text Mining Provider</td>\n",
       "      <td>CORD Gene API</td>\n",
       "      <td>None</td>\n",
       "      <td>ChemicalSubstance</td>\n",
       "      <td>CHEBI:46756</td>\n",
       "      <td>CHEBI:CHEBI:46756</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        input input_type       pred1 pred1_source  \\\n",
       "16504  HYPERPHENYLALANINAEMIA    Disease   caused_by       SEMMED   \n",
       "8209   HYPERPHENYLALANINAEMIA    Disease  related_to       SEMMED   \n",
       "31798  HYPERPHENYLALANINAEMIA    Disease   caused_by       SEMMED   \n",
       "1929   HYPERPHENYLALANINAEMIA    Disease   caused_by       SEMMED   \n",
       "10532  HYPERPHENYLALANINAEMIA    Disease   caused_by       SEMMED   \n",
       "35512  HYPERPHENYLALANINAEMIA    Disease   caused_by       SEMMED   \n",
       "40287  HYPERPHENYLALANINAEMIA    Disease  treated_by       SEMMED   \n",
       "40027  HYPERPHENYLALANINAEMIA    Disease   caused_by       SEMMED   \n",
       "22054  HYPERPHENYLALANINAEMIA    Disease   caused_by       SEMMED   \n",
       "45089  HYPERPHENYLALANINAEMIA    Disease   caused_by       SEMMED   \n",
       "\n",
       "                pred1_api                       pred1_pubmed node1_type  \\\n",
       "16504  SEMMED Disease API                           10727395       Gene   \n",
       "8209   SEMMED Disease API                            9675155       Gene   \n",
       "31798  SEMMED Disease API                    9222755,9667588       Gene   \n",
       "1929   SEMMED Disease API                           10874306       Gene   \n",
       "10532  SEMMED Disease API                             699146       Gene   \n",
       "35512  SEMMED Disease API                           10727395       Gene   \n",
       "40287  SEMMED Disease API                            8841415       Gene   \n",
       "40027  SEMMED Disease API                           15133828       Gene   \n",
       "22054  SEMMED Disease API                           28794131       Gene   \n",
       "45089  SEMMED Disease API  1301201,8892014,10495930,10495930       Gene   \n",
       "\n",
       "      node1_name        node1_id                    pred2  \\\n",
       "16504   C0014442   UMLS:C0014442  negatively_regulated_by   \n",
       "8209    C0017337   UMLS:C0017337  positively_regulated_by   \n",
       "31798   C0017337   UMLS:C0017337  positively_regulated_by   \n",
       "1929         PTS   NCBIGene:5805               related_to   \n",
       "10532   C0030016   UMLS:C0030016  positively_regulated_by   \n",
       "35512   C0014442   UMLS:C0014442     positively_regulates   \n",
       "40287        PTS   NCBIGene:5805               related_to   \n",
       "40027    AKR1B10  NCBIGene:57016     negatively_regulates   \n",
       "22054      HSPA4   NCBIGene:3308  positively_regulated_by   \n",
       "45089        PAH   NCBIGene:5053               related_to   \n",
       "\n",
       "                          pred2_source                    pred2_api  \\\n",
       "16504                           SEMMED              SEMMED Gene API   \n",
       "8209                            SEMMED              SEMMED Gene API   \n",
       "31798                           SEMMED              SEMMED Gene API   \n",
       "1929   Translator Text Mining Provider                CORD Gene API   \n",
       "10532                           SEMMED              SEMMED Gene API   \n",
       "35512                           SEMMED              SEMMED Gene API   \n",
       "40287                         scigraph  Automat CORD19 Scigraph API   \n",
       "40027                           SEMMED              SEMMED Gene API   \n",
       "22054                           SEMMED              SEMMED Gene API   \n",
       "45089  Translator Text Mining Provider                CORD Gene API   \n",
       "\n",
       "                     pred2_pubmed        output_type             output_name  \\\n",
       "16504           11839529,19443747  ChemicalSubstance                  U 0126   \n",
       "8209            21984199,26086415  ChemicalSubstance              ISOFLURANE   \n",
       "31798  17684401,18412988,17693008  ChemicalSubstance           MICRONUTRIENT   \n",
       "1929                         None  ChemicalSubstance  .BETA.-D-GLUCOPYRANOSE   \n",
       "10532                    11731036  ChemicalSubstance               GENISTEIN   \n",
       "35512    4291334,7358701,11980911  ChemicalSubstance            DEOXYCHOLATE   \n",
       "40287                        None  ChemicalSubstance            CHLORAMBUCIL   \n",
       "40027                    28322780  ChemicalSubstance        TIAPROFENIC ACID   \n",
       "22054                    26307981  ChemicalSubstance           MALONALDEHYDE   \n",
       "45089                        None  ChemicalSubstance             CHEBI:46756   \n",
       "\n",
       "                         output_id  \n",
       "16504                  name:U 0126  \n",
       "8209               name:ISOFLURANE  \n",
       "31798           name:MICRONUTRIENT  \n",
       "1929   name:.BETA.-D-GLUCOPYRANOSE  \n",
       "10532               name:GENISTEIN  \n",
       "35512            name:DEOXYCHOLATE  \n",
       "40287            name:CHLORAMBUCIL  \n",
       "40027        name:TIAPROFENIC ACID  \n",
       "22054           name:MALONALDEHYDE  \n",
       "45089            CHEBI:CHEBI:46756  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = fc.display_table_view()\n",
    "\n",
    "# because UMLS is not currently well-integrated in our ID-to-object translation system, removing UMLS-only outputs here\n",
    "patternDel = \"^UMLS:C\\d+\"\n",
    "filter = df.output_id.str.contains(patternDel)\n",
    "df = df[~filter]\n",
    "\n",
    "df.sample(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "While most results are based on edges from [semmed](https://skr3.nlm.nih.gov/SemMed/), edges from [DGIdb](http://www.dgidb.org/), [biolink](https://monarchinitiative.org/), [disgenet](http://www.disgenet.org/), and [mychem.info](https://mychem.info) were also retrieved from their respective APIs. \n",
    "\n",
    "Next, let's examine which chemical compounds were most frequently mentioned."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "PHARMACEUTICAL PREPARATIONS    122\n",
       "HYDROGEN PEROXIDE              119\n",
       ".BETA.-D-GLUCOPYRANOSE         116\n",
       "NADH                           105\n",
       "NITRIC OXIDE                   101\n",
       "BINDING SITE                    97\n",
       "3-METHYLADENINE DNA             94\n",
       "CHOLESTEROL                     94\n",
       "OXYGEN SPECIES, REACTIVE        92\n",
       "ALCOHOL                         91\n",
       "Name: output_name, dtype: int64"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.output_name.value_counts().head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Hyperphenylalaninemia is a condition characterized by elevated levels of phenylalanine in the blood. This phenotype is strongly associated with phenylketonuria (PKU), an inherited, genetic disorder that affects the ability to metabolize phenylalanine.  **Sapropterin** is a naturally-occurring cofactor associated with several enzymatic processes, including the metabolism of phenylalanine to tyrosine. It has been [FDA-approved](https://www.centerwatch.com/drug-information/fda-approved-drugs/drug/972/kuvan-sapropterin-dihydrochloride) as a treatment for PKU patients.  Tyrosine is also a precursor to several neurotransmitters, including **norepinephrine** and **dopamine**.\n",
    "\n",
    "While several of the listed compounds would be considered positive controls, others on the list could be viewed as **testable hypotheses and discovery opportunities** to be evaluated by domain experts. For example, let's examine which genes link hyperphenylalaninemia to nitric oxide."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>input</th>\n",
       "      <th>input_type</th>\n",
       "      <th>pred1</th>\n",
       "      <th>pred1_source</th>\n",
       "      <th>pred1_api</th>\n",
       "      <th>pred1_pubmed</th>\n",
       "      <th>node1_type</th>\n",
       "      <th>node1_name</th>\n",
       "      <th>node1_id</th>\n",
       "      <th>pred2</th>\n",
       "      <th>pred2_source</th>\n",
       "      <th>pred2_api</th>\n",
       "      <th>pred2_pubmed</th>\n",
       "      <th>output_type</th>\n",
       "      <th>output_name</th>\n",
       "      <th>output_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>2318</td>\n",
       "      <td>HYPERPHENYLALANINAEMIA</td>\n",
       "      <td>Disease</td>\n",
       "      <td>treated_by</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Disease API</td>\n",
       "      <td>25261586,25261586</td>\n",
       "      <td>Gene</td>\n",
       "      <td>C1710338</td>\n",
       "      <td>UMLS:C1710338</td>\n",
       "      <td>negatively_regulates</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Gene API</td>\n",
       "      <td>23769841</td>\n",
       "      <td>ChemicalSubstance</td>\n",
       "      <td>NITRIC OXIDE</td>\n",
       "      <td>name:NITRIC OXIDE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2319</td>\n",
       "      <td>HYPERPHENYLALANINAEMIA</td>\n",
       "      <td>Disease</td>\n",
       "      <td>treated_by</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Disease API</td>\n",
       "      <td>25261586,25261586</td>\n",
       "      <td>Gene</td>\n",
       "      <td>C1710338</td>\n",
       "      <td>UMLS:C1710338</td>\n",
       "      <td>negatively_regulates</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Gene API</td>\n",
       "      <td>15265010</td>\n",
       "      <td>ChemicalSubstance</td>\n",
       "      <td>NITRIC OXIDE</td>\n",
       "      <td>name:NITRIC OXIDE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2320</td>\n",
       "      <td>HYPERPHENYLALANINAEMIA</td>\n",
       "      <td>Disease</td>\n",
       "      <td>treated_by</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Disease API</td>\n",
       "      <td>25261586,25261586</td>\n",
       "      <td>Gene</td>\n",
       "      <td>C1710338</td>\n",
       "      <td>UMLS:C1710338</td>\n",
       "      <td>physically_interacts_with</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Gene API</td>\n",
       "      <td>25049660,23072349</td>\n",
       "      <td>ChemicalSubstance</td>\n",
       "      <td>NITRIC OXIDE</td>\n",
       "      <td>name:NITRIC OXIDE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2321</td>\n",
       "      <td>HYPERPHENYLALANINAEMIA</td>\n",
       "      <td>Disease</td>\n",
       "      <td>treated_by</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Disease API</td>\n",
       "      <td>25261586,25261586</td>\n",
       "      <td>Gene</td>\n",
       "      <td>C1710338</td>\n",
       "      <td>UMLS:C1710338</td>\n",
       "      <td>physically_interacts_with</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Gene API</td>\n",
       "      <td>23806552,16831427</td>\n",
       "      <td>ChemicalSubstance</td>\n",
       "      <td>NITRIC OXIDE</td>\n",
       "      <td>name:NITRIC OXIDE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2322</td>\n",
       "      <td>HYPERPHENYLALANINAEMIA</td>\n",
       "      <td>Disease</td>\n",
       "      <td>treated_by</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Disease API</td>\n",
       "      <td>25261586,25261586</td>\n",
       "      <td>Gene</td>\n",
       "      <td>C1710338</td>\n",
       "      <td>UMLS:C1710338</td>\n",
       "      <td>positively_regulated_by</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Gene API</td>\n",
       "      <td>10952392</td>\n",
       "      <td>ChemicalSubstance</td>\n",
       "      <td>NITRIC OXIDE</td>\n",
       "      <td>name:NITRIC OXIDE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2422</td>\n",
       "      <td>HYPERPHENYLALANINAEMIA</td>\n",
       "      <td>Disease</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Disease API</td>\n",
       "      <td>28794131</td>\n",
       "      <td>Gene</td>\n",
       "      <td>HSPA4</td>\n",
       "      <td>NCBIGene:3308</td>\n",
       "      <td>positively_regulated_by</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Gene API</td>\n",
       "      <td>7544743,9085348,9085348,9276725,10805410,10805...</td>\n",
       "      <td>ChemicalSubstance</td>\n",
       "      <td>NITRIC OXIDE</td>\n",
       "      <td>name:NITRIC OXIDE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2423</td>\n",
       "      <td>HYPERPHENYLALANINAEMIA</td>\n",
       "      <td>Disease</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Disease API</td>\n",
       "      <td>28794131</td>\n",
       "      <td>Gene</td>\n",
       "      <td>HSPA4</td>\n",
       "      <td>NCBIGene:3308</td>\n",
       "      <td>positively_regulates</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Gene API</td>\n",
       "      <td>11374865</td>\n",
       "      <td>ChemicalSubstance</td>\n",
       "      <td>NITRIC OXIDE</td>\n",
       "      <td>name:NITRIC OXIDE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2424</td>\n",
       "      <td>HYPERPHENYLALANINAEMIA</td>\n",
       "      <td>Disease</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Disease API</td>\n",
       "      <td>7766952</td>\n",
       "      <td>Gene</td>\n",
       "      <td>CACNA1A</td>\n",
       "      <td>NCBIGene:773</td>\n",
       "      <td>negatively_regulates</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Gene API</td>\n",
       "      <td>23949212</td>\n",
       "      <td>ChemicalSubstance</td>\n",
       "      <td>NITRIC OXIDE</td>\n",
       "      <td>name:NITRIC OXIDE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2425</td>\n",
       "      <td>HYPERPHENYLALANINAEMIA</td>\n",
       "      <td>Disease</td>\n",
       "      <td>related_to</td>\n",
       "      <td>scibite</td>\n",
       "      <td>Automat CORD19 Scibite API</td>\n",
       "      <td>None</td>\n",
       "      <td>Gene</td>\n",
       "      <td>GCHFR</td>\n",
       "      <td>NCBIGene:2644</td>\n",
       "      <td>physically_interacts_with</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Gene API</td>\n",
       "      <td>18372436,18372436,18372436</td>\n",
       "      <td>ChemicalSubstance</td>\n",
       "      <td>NITRIC OXIDE</td>\n",
       "      <td>name:NITRIC OXIDE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2426</td>\n",
       "      <td>HYPERPHENYLALANINAEMIA</td>\n",
       "      <td>Disease</td>\n",
       "      <td>related_to</td>\n",
       "      <td>scibite</td>\n",
       "      <td>Automat CORD19 Scibite API</td>\n",
       "      <td>None</td>\n",
       "      <td>Gene</td>\n",
       "      <td>GCHFR</td>\n",
       "      <td>NCBIGene:2644</td>\n",
       "      <td>physically_interacts_with</td>\n",
       "      <td>SEMMED</td>\n",
       "      <td>SEMMED Gene API</td>\n",
       "      <td>21693765</td>\n",
       "      <td>ChemicalSubstance</td>\n",
       "      <td>NITRIC OXIDE</td>\n",
       "      <td>name:NITRIC OXIDE</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>101 rows  16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       input input_type       pred1 pred1_source  \\\n",
       "2318  HYPERPHENYLALANINAEMIA    Disease  treated_by       SEMMED   \n",
       "2319  HYPERPHENYLALANINAEMIA    Disease  treated_by       SEMMED   \n",
       "2320  HYPERPHENYLALANINAEMIA    Disease  treated_by       SEMMED   \n",
       "2321  HYPERPHENYLALANINAEMIA    Disease  treated_by       SEMMED   \n",
       "2322  HYPERPHENYLALANINAEMIA    Disease  treated_by       SEMMED   \n",
       "...                      ...        ...         ...          ...   \n",
       "2422  HYPERPHENYLALANINAEMIA    Disease   caused_by       SEMMED   \n",
       "2423  HYPERPHENYLALANINAEMIA    Disease   caused_by       SEMMED   \n",
       "2424  HYPERPHENYLALANINAEMIA    Disease   caused_by       SEMMED   \n",
       "2425  HYPERPHENYLALANINAEMIA    Disease  related_to      scibite   \n",
       "2426  HYPERPHENYLALANINAEMIA    Disease  related_to      scibite   \n",
       "\n",
       "                       pred1_api       pred1_pubmed node1_type node1_name  \\\n",
       "2318          SEMMED Disease API  25261586,25261586       Gene   C1710338   \n",
       "2319          SEMMED Disease API  25261586,25261586       Gene   C1710338   \n",
       "2320          SEMMED Disease API  25261586,25261586       Gene   C1710338   \n",
       "2321          SEMMED Disease API  25261586,25261586       Gene   C1710338   \n",
       "2322          SEMMED Disease API  25261586,25261586       Gene   C1710338   \n",
       "...                          ...                ...        ...        ...   \n",
       "2422          SEMMED Disease API           28794131       Gene      HSPA4   \n",
       "2423          SEMMED Disease API           28794131       Gene      HSPA4   \n",
       "2424          SEMMED Disease API            7766952       Gene    CACNA1A   \n",
       "2425  Automat CORD19 Scibite API               None       Gene      GCHFR   \n",
       "2426  Automat CORD19 Scibite API               None       Gene      GCHFR   \n",
       "\n",
       "           node1_id                      pred2 pred2_source        pred2_api  \\\n",
       "2318  UMLS:C1710338       negatively_regulates       SEMMED  SEMMED Gene API   \n",
       "2319  UMLS:C1710338       negatively_regulates       SEMMED  SEMMED Gene API   \n",
       "2320  UMLS:C1710338  physically_interacts_with       SEMMED  SEMMED Gene API   \n",
       "2321  UMLS:C1710338  physically_interacts_with       SEMMED  SEMMED Gene API   \n",
       "2322  UMLS:C1710338    positively_regulated_by       SEMMED  SEMMED Gene API   \n",
       "...             ...                        ...          ...              ...   \n",
       "2422  NCBIGene:3308    positively_regulated_by       SEMMED  SEMMED Gene API   \n",
       "2423  NCBIGene:3308       positively_regulates       SEMMED  SEMMED Gene API   \n",
       "2424   NCBIGene:773       negatively_regulates       SEMMED  SEMMED Gene API   \n",
       "2425  NCBIGene:2644  physically_interacts_with       SEMMED  SEMMED Gene API   \n",
       "2426  NCBIGene:2644  physically_interacts_with       SEMMED  SEMMED Gene API   \n",
       "\n",
       "                                           pred2_pubmed        output_type  \\\n",
       "2318                                           23769841  ChemicalSubstance   \n",
       "2319                                           15265010  ChemicalSubstance   \n",
       "2320                                  25049660,23072349  ChemicalSubstance   \n",
       "2321                                  23806552,16831427  ChemicalSubstance   \n",
       "2322                                           10952392  ChemicalSubstance   \n",
       "...                                                 ...                ...   \n",
       "2422  7544743,9085348,9085348,9276725,10805410,10805...  ChemicalSubstance   \n",
       "2423                                           11374865  ChemicalSubstance   \n",
       "2424                                           23949212  ChemicalSubstance   \n",
       "2425                         18372436,18372436,18372436  ChemicalSubstance   \n",
       "2426                                           21693765  ChemicalSubstance   \n",
       "\n",
       "       output_name          output_id  \n",
       "2318  NITRIC OXIDE  name:NITRIC OXIDE  \n",
       "2319  NITRIC OXIDE  name:NITRIC OXIDE  \n",
       "2320  NITRIC OXIDE  name:NITRIC OXIDE  \n",
       "2321  NITRIC OXIDE  name:NITRIC OXIDE  \n",
       "2322  NITRIC OXIDE  name:NITRIC OXIDE  \n",
       "...            ...                ...  \n",
       "2422  NITRIC OXIDE  name:NITRIC OXIDE  \n",
       "2423  NITRIC OXIDE  name:NITRIC OXIDE  \n",
       "2424  NITRIC OXIDE  name:NITRIC OXIDE  \n",
       "2425  NITRIC OXIDE  name:NITRIC OXIDE  \n",
       "2426  NITRIC OXIDE  name:NITRIC OXIDE  \n",
       "\n",
       "[101 rows x 16 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df.output_name==\"NITRIC OXIDE\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "BTE identifies the genes [HSPA4](https://www.ncbi.nlm.nih.gov/gene/3308) and [GCH1](https://www.ncbi.nlm.nih.gov/gene/2643) as being the key nodes that join hyperphenylalaninemia with nitric oxide."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Conclusions and caveats\n",
    "\n",
    "This notebook demonstrated the use of BioThings Explorer in PREDICDT mode to identify chemical compounds related to hyperphenylalaninemia. BioThings Explorer autonomously queried a **distributed knowledge graph of biomedical APIs** to find the most common compounds, and several of the top hits had a clear biochemical and clinical relevance to hyperphenylalaninemia.\n",
    "\n",
    "There are still many areas for improvement (and some areas in which BioThings Explorer is still buggy).  And of course, BioThings Explorer is dependent on the accessibility of the APIs that comprise the distributed knowledge graph.  Nevertheless, we encourage users to try other variants of the PREDICT queries demonstrated in this notebook."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
